{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Debug Notebook\n",
    "\n",
    "This notebook was used to debug the annotation that is based on the GFF3 file directly (without creating the individual json files), which is now available in the file `Utils/Gene_annotation.py`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Global configuration (common to all projects).\n",
    "'''\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os, sys, re\n",
    "import pathlib\n",
    "import pickle\n",
    "import matplotlib.pyplot as plt\n",
    "import glob\n",
    "# import subprocess\n",
    "\n",
    "# make sure PYTHONPATH is set, containing the Utils folder.\n",
    "import mysys_init as si"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ensembl_gene_annotations_utils as egna\n",
    "from dataclasses import dataclass, field\n",
    "\n",
    "Ensembl_release: str = \"109\"\n",
    "# --------------------------\n",
    "\n",
    "\n",
    "GFF3_file = pathlib.Path(f\"/Users/yoramzarai/work/mystuff/Ramot/Projects/Cancer_mut/Data/Ensembl_gene_annotation/Homo_sapiens.GRCh38.{Ensembl_release}.gff3\") if sys.platform == \"darwin\" \\\n",
    "    else pathlib.Path(f\"/tamir2/yoramzar/Projects/Cancer_mut/Data/Ensembl_gene_annotation/Homo_sapiens.GRCh38.{Ensembl_release}.gff3\")\n",
    "\n",
    "\n",
    "def ensembl_gff3_df(file: pathlib.Path = GFF3_file) -> pd.DataFrame:\n",
    "     \"\"\"For users of gene annotation.\"\"\"\n",
    "     return egna.load_ensembl_human_gff3_annotation_file(file)\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class Transcript_gff3_cls:\n",
    "    # dataframe (result of pd.read_csv(GFF3 file)) or a path to an Ensembl GFF3 file\n",
    "    gene: str\n",
    "    gff3_data: pathlib.Path | pd.DataFrame\n",
    "    verbose: bool = True  # set to False at instantiation to suppress prints\n",
    "\n",
    "    # these are set by the __post_init__ method.\n",
    "    gene_ID: str = ''\n",
    "    chrm: str = ''\n",
    "    rev: bool = False\n",
    "    gene_start: int = 0\n",
    "    gene_end: int = 0\n",
    "    gene_type: str = ''\n",
    "    gene_desc: str = ''\n",
    "    gene_ver: int = 0\n",
    "    transcripts: list = None \n",
    "\n",
    "    source: str = ''\n",
    "\n",
    "    transcripts_info: dict = None  # keys are transcripts, values are transcript details\n",
    "    exon_intron_maps: dict = None  # keys are transcripts, values are maps in df format\n",
    "\n",
    "    __protein_coding_labels_in_biotype: list = None\n",
    "\n",
    "    def __post_init__(self) -> None:\n",
    "        if type(self.gff3_data) is not pd.DataFrame:\n",
    "            self.source = str(self.gff3_data)\n",
    "            if self.verbose: print(f\"Loading {self.gff3_data} to a dataframe...\", end='')\n",
    "            self.gff3_data = egna.load_ensembl_human_gff3_annotation_file(self.gff3_data)\n",
    "            if self.verbose: print(\"Done.\")\n",
    "        else:\n",
    "             self.source = f'Input DataFrame (ID={id(self.gff3_data)}, size={self.gff3_data.shape}).'\n",
    "        \n",
    "        # extract gene and generate the description dictionary\n",
    "        if not (d := egna.extract_gene_dict(self.gff3_data, self.gene)):\n",
    "             raise ValueError(f\"Error in extracting the gene {self.gene} from GFF3 dataframe !! Exiting.\")\n",
    "\n",
    "        self.chrm = d['chrm']\n",
    "        self.gene_start = d['gene_start']\n",
    "        self.gene_end = d['gene_end']\n",
    "        self.rev = d['rev']\n",
    "        self.gene_ID = d['gene_ID']\n",
    "        self.gene_type = d['gene_type']\n",
    "        self.gene_desc = d['gene_desc']\n",
    "        self.gene_ver = d['gene_ver']\n",
    "        self.transcripts = d['transcripts']\n",
    "\n",
    "        if self.__protein_coding_labels_in_biotype is None:\n",
    "                # see https://www.ensembl.org/info/genome/genebuild/biotypes.html\n",
    "                self.__protein_coding_labels_in_biotype = [\n",
    "                    \"nonsense_mediated_decay\",\n",
    "                    \"protein_coding\",\n",
    "                ]\n",
    "\n",
    "        # creates transcript information for each transcript. This populates self.transcripts_info\n",
    "        #self.__gen_transcripts_info()\n",
    "\n",
    "        # creates exon-intron map for each transcript. This populates self.exon_intron_maps\n",
    "        #self.__create_exon_intron_map()\n",
    "\n",
    "        \n",
    "class Gene_cls(Transcript_gff3_cls):  # CHANGE THIS TO Gene_gff3_cls\n",
    "    def __init__(self, gene: str, gff3_file: pathlib.Path = GFF3_file, verbose: bool = True) -> None:\n",
    "        self.gene = gene\n",
    "        self.gff3_data = gff3_file\n",
    "        self.verbose = verbose\n",
    "        \n",
    "\n",
    "        super().__post_init__()\n",
    "     \n",
    "\n",
    "    def __repr__(self) -> str:\n",
    "        return f\"Gene_cls('{self.gene}')\"\n",
    "\n",
    "    def __str__(self) -> str:\n",
    "        return f\"Gene annotation class based on gene name.\\n\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading GFF3 file (for testing)\n",
    "gff3_df = ensembl_gff3_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# test\n",
    "gene: str = 'IDH1' #'ENSG00000267318'# 'PIK3CA'\n",
    "# ================\n",
    "#a = Transcript_gff3_cls(gene, gff3_df)\n",
    "#a = Transcript_gff3_cls(gene, GFF3_file, verbose=False)\n",
    "a = Gene_cls(gene)\n",
    "\n",
    "print(a.source, a.gene, a.gene_ID, a.chrm, a.gene_start, a.gene_end, a.rev, a.gene_type, a.gene_desc, a.gene_ver, len(a.transcripts), sep='\\n')\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test the (new) API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Gene_annotation as gaut"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading GFF3 file (for testing)\n",
    "gff3_df = gaut.ensembl_gff3_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrm</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1</td>\n",
       "      <td>248956422</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469</td>\n",
       "      <td>11240</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650</td>\n",
       "      <td>10657</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655</td>\n",
       "      <td>10657</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678</td>\n",
       "      <td>10687</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3414006</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>26626966</td>\n",
       "      <td>26627137</td>\n",
       "      <td>0.994</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=rank %3D 1;logic_name=firstef</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3414007</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>26627457</td>\n",
       "      <td>26628186</td>\n",
       "      <td>0.997</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=rank %3D 1;logic_name=firstef</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3414008</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>56855244</td>\n",
       "      <td>56855488</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000235857;Name=CTBP2P1;biotype=p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3414009</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>pseudogenic_transcript</td>\n",
       "      <td>56855244</td>\n",
       "      <td>56855488</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000431853;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3414010</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>56855244</td>\n",
       "      <td>56855488</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000431853;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3414011 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chrm  Source                    Type     Start        End    Score  \\\n",
       "0          1  GRCh38              chromosome         1  248956422        .   \n",
       "1          1       .       biological_region     10469      11240  1.3e+03   \n",
       "2          1       .       biological_region     10650      10657    0.999   \n",
       "3          1       .       biological_region     10655      10657    0.999   \n",
       "4          1       .       biological_region     10678      10687    0.999   \n",
       "...      ...     ...                     ...       ...        ...      ...   \n",
       "3414006    Y       .       biological_region  26626966   26627137    0.994   \n",
       "3414007    Y       .       biological_region  26627457   26628186    0.997   \n",
       "3414008    Y  havana              pseudogene  56855244   56855488        .   \n",
       "3414009    Y  havana  pseudogenic_transcript  56855244   56855488        .   \n",
       "3414010    Y  havana                    exon  56855244   56855488        .   \n",
       "\n",
       "        Strand Phase                                         Attributes  \n",
       "0            .     .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "1            .     .           external_name=oe %3D 0.79;logic_name=cpg  \n",
       "2            +     .                                 logic_name=eponine  \n",
       "3            -     .                                 logic_name=eponine  \n",
       "4            +     .                                 logic_name=eponine  \n",
       "...        ...   ...                                                ...  \n",
       "3414006      -     .        external_name=rank %3D 1;logic_name=firstef  \n",
       "3414007      +     .        external_name=rank %3D 1;logic_name=firstef  \n",
       "3414008      +     .  ID=gene:ENSG00000235857;Name=CTBP2P1;biotype=p...  \n",
       "3414009      +     .  ID=transcript:ENST00000431853;Parent=gene:ENSG...  \n",
       "3414010      +     .  Parent=transcript:ENST00000431853;Name=ENSE000...  \n",
       "\n",
       "[3414011 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrm</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000186092;Name=OR4F5;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000284733;Name=OR4F29;biotype=pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>502</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000284662;Name=OR4F16;biotype=pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1097</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>923923</td>\n",
       "      <td>944575</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000187634;Name=SAMD11;biotype=pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1459</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>944203</td>\n",
       "      <td>959309</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000188976;Name=NOC2L;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3412977</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>24607560</td>\n",
       "      <td>24639207</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000183795;Name=BPY2B;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3413043</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>24763069</td>\n",
       "      <td>24813492</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000187191;Name=DAZ3;biotype=prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3413169</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>24833843</td>\n",
       "      <td>24907040</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000205916;Name=DAZ4;biotype=prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3413567</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>25030901</td>\n",
       "      <td>25062548</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000185894;Name=BPY2C;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3413759</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>25622117</td>\n",
       "      <td>25624902</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000172288;Name=CDY1;biotype=prot...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21531 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chrm          Source  Type     Start       End Score Strand Phase  \\\n",
       "84         1  ensembl_havana  gene     65419     71585     .      +     .   \n",
       "370        1  ensembl_havana  gene    450740    451678     .      -     .   \n",
       "502        1  ensembl_havana  gene    685716    686654     .      -     .   \n",
       "1097       1  ensembl_havana  gene    923923    944575     .      +     .   \n",
       "1459       1  ensembl_havana  gene    944203    959309     .      -     .   \n",
       "...      ...             ...   ...       ...       ...   ...    ...   ...   \n",
       "3412977    Y  ensembl_havana  gene  24607560  24639207     .      +     .   \n",
       "3413043    Y  ensembl_havana  gene  24763069  24813492     .      -     .   \n",
       "3413169    Y  ensembl_havana  gene  24833843  24907040     .      +     .   \n",
       "3413567    Y  ensembl_havana  gene  25030901  25062548     .      -     .   \n",
       "3413759    Y  ensembl_havana  gene  25622117  25624902     .      +     .   \n",
       "\n",
       "                                                Attributes  \n",
       "84       ID=gene:ENSG00000186092;Name=OR4F5;biotype=pro...  \n",
       "370      ID=gene:ENSG00000284733;Name=OR4F29;biotype=pr...  \n",
       "502      ID=gene:ENSG00000284662;Name=OR4F16;biotype=pr...  \n",
       "1097     ID=gene:ENSG00000187634;Name=SAMD11;biotype=pr...  \n",
       "1459     ID=gene:ENSG00000188976;Name=NOC2L;biotype=pro...  \n",
       "...                                                    ...  \n",
       "3412977  ID=gene:ENSG00000183795;Name=BPY2B;biotype=pro...  \n",
       "3413043  ID=gene:ENSG00000187191;Name=DAZ3;biotype=prot...  \n",
       "3413169  ID=gene:ENSG00000205916;Name=DAZ4;biotype=prot...  \n",
       "3413567  ID=gene:ENSG00000185894;Name=BPY2C;biotype=pro...  \n",
       "3413759  ID=gene:ENSG00000172288;Name=CDY1;biotype=prot...  \n",
       "\n",
       "[21531 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID=gene:ENSG00000186092;Name=OR4F5;biotype=protein_coding;description=olfactory receptor family 4 subfamily F member 5 [Source:HGNC Symbol%3BAcc:HGNC:14825];gene_id=ENSG00000186092;logic_name=ensembl_havana_gene_homo_sapiens;version=7\n"
     ]
    }
   ],
   "source": [
    "display(gff3_df)\n",
    "d = gff3_df.query(\"Type == 'gene'\")\n",
    "display(d)\n",
    "print(d.iloc[0]['Attributes'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene: str = 'IDH1' #'ENSG00000267318'# 'PIK3CA'\n",
    "# ================\n",
    "#a = gaut.Transcript_gff3_cls(gene, gff3_df)\n",
    "#a = gaut.Transcript_gff3_cls(gene, gaut.GFF3_file, verbose=False)\n",
    "a = gaut.Gene_cls(gene)\n",
    "\n",
    "print(a.source, a.gene, a.gene_ID, a.chrm, a.gene_start, a.gene_end, a.rev, a.gene_type, a.gene_desc, a.gene_ver, len(a.transcripts), sep='\\n')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv311",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
