{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene Annotation Usage Example\n",
    "This notebook provides gene annotation examples, using the `Gene_cls` or `Gene_gff3_cls` classes from `Gene_annotation.py`.\n",
    "\n",
    "The underlying data is the Ensembl GFF3 annotation file from [Ensembl gene annotation GFF3 file](https://useast.ensembl.org/Homo_sapiens/Info/Index).\n",
    "\n",
    "Requirements: Python >= 3.10, Pandas, and Numpy. In addition, add \n",
    "\n",
    "`export PYTHONPATH=\"$PYTHONPATH:/tamir2/yoramzar/Projects/Cancer_mut/Utils\"` \n",
    "\n",
    "to your `.bashrc`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Global configuration.\n",
    "'''\n",
    "import pathlib\n",
    "import pandas as pd\n",
    "\n",
    "# make sure PYTHONPATH is set, containing the Utils folder.\n",
    "import mysys_init as si\n",
    "\n",
    "# import API\n",
    "import Gene_annotation as gaut\n",
    "\n",
    "# set folder to create reports\n",
    "report_path = pathlib.Path(si.ROOT_path) / 'Ensembl_gene_anot' / 'Reports'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example - by Gene Name\n",
    "\n",
    "We use the gene name to instantiate a `Gene_cls` object, which provides the gene's annotation.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /Users/yoramzarai/work/mystuff/Ramot/Projects/Cancer_mut/Data/Ensembl_gene_annotation/Homo_sapiens.GRCh38.110.gff3.gz to a dataframe...Done.\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Select a gene and a corresponding transcript.\n",
    "'''\n",
    "#gene, transcript_id = 'EGFR', 'ENST00000275493'\n",
    "#gene, transcript_id = 'IDH1', 'ENST00000345146'\n",
    "gene, transcript_id = 'PIK3CA', 'ENST00000263967'\n",
    "\n",
    "#gene, transcript_id = 'ENSG00000290826', 'ENST00000642116'  # example of a ncRNA gene ()\n",
    "\n",
    "#gene, transcript_id = 'PIK3C2A', 'ENST00000533645'\n",
    "#gene, transcript_id = 'KLF16', 'ENST00000250916'\n",
    "\n",
    "#gene, transcript_id = 'ENSG00000290149', 'ENST00000476620'\n",
    "# ---------------------------------------------------------\n",
    "\n",
    "gene_obj = gaut.Gene_cls(gene)\n",
    "#gene_obj = gaut.Gene_cls(gene, verbose=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "gene='PIK3CA' information:\n",
      "Gene name=PIK3CA, Gene ID=ENSG00000121879\n",
      "Source=/Users/yoramzarai/work/mystuff/Ramot/Projects/Cancer_mut/Data/Ensembl_gene_annotation/Homo_sapiens.GRCh38.110.gff3.gz\n",
      "Positive strand=True.\n",
      "Type=protein_coding, version=6.\n",
      "Description=phosphatidylinositol-4%2C5-bisphosphate 3-kinase catalytic subunit alpha [Source:HGNC Symbol%3BAcc:HGNC:8975]\n",
      "Gene region=chr3:179,148,114-179,240,093.\n",
      "10 transcripts.\n",
      "\n",
      "PIK3CA contains 10 transcripts.\n",
      "PIK3CA transcript IDs:\n",
      "1. ENST00000477735\n",
      "2. ENST00000263967\n",
      "3. ENST00000643187\n",
      "4. ENST00000675786\n",
      "5. ENST00000468036\n",
      "6. ENST00000675467\n",
      "7. ENST00000674534\n",
      "8. ENST00000674622\n",
      "9. ENST00000462255\n",
      "10. ENST00000675796\n",
      "\n",
      "Gene start = 179,148,114, gene end = 179,240,093\n",
      "\n",
      "transcript_id='ENST00000263967': start site (TSS) = 179,148,357, end site (TES) = 179,240,093.\n",
      "\n",
      "First bp of the Start codon located in chr3:179,198,826, and is bp #324 in the RNA.\n",
      "First bp of the Stop codon located in chr3:179,234,362, and is bp #3528 in the RNA.\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "General information\n",
    "\"\"\"\n",
    "print(f\"\\n{gene=} information:\")\n",
    "gene_obj.info()\n",
    "\n",
    "print(f\"\\n{gene} contains {len(gene_obj)} transcripts.\")\n",
    "gene_obj.show_all_transcript_IDs()\n",
    "\n",
    "print(f\"\\nGene start = {gene_obj.gene_start:,}, gene end = {gene_obj.gene_end:,}\")\n",
    "\n",
    "(ts, te) = gene_obj.transcript_boundaries(transcript_id)\n",
    "print(f\"\\n{transcript_id=}: start site (TSS) = {ts:,}, end site (TES) = {te:,}.\")\n",
    "\n",
    "# start and stop codons positions\n",
    "if (codons_info := gene_obj.start_and_stop_codons_pos(transcript_id)) is not None:\n",
    "    print(f\"\\nFirst bp of the Start codon located in chr{gene_obj.chrm}:{codons_info[0][0]:,}, and is bp #{codons_info[0][1]} in the RNA.\")\n",
    "    print(f\"First bp of the Stop codon located in chr{gene_obj.chrm}:{codons_info[1][0]:,}, and is bp #{codons_info[1][1]} in the RNA.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detailed transcript_id='ENST00000263967' information:\n",
      "ENST00000263967 information:\n",
      "Name=PIK3CA-201, ID=ENST00000263967.4, version=4\n",
      "Type=mRNA, biotype=protein_coding\n",
      "Start=179,148,357, end=179,240,093\n",
      "protein ID = ENSP00000263967\n",
      "ccdsid=CCDS43171.1\n",
      "ORF start offset = 323, ORF end offset = 3,529 (0-based offsets from the beginning of the RNA)\n",
      "ORF start chromosome position = 179,198,826, ORF end chromosome position = 179,234,364\n",
      "Exon regions:\n",
      "\t1. 179,148,357 - 179,148,603, ID=ENSE00001493081, phase=-1, end phase=-1\n",
      "\t2. 179,198,750 - 179,199,177, ID=ENSE00001139995, phase=-1, end phase=1\n",
      "\t3. 179,199,690 - 179,199,899, ID=ENSE00000997375, phase=1, end phase=1\n",
      "\t4. 179,201,290 - 179,201,540, ID=ENSE00001077693, phase=1, end phase=0\n",
      "\t5. 179,203,544 - 179,203,789, ID=ENSE00001077692, phase=0, end phase=0\n",
      "\t6. 179,204,503 - 179,204,588, ID=ENSE00001077694, phase=0, end phase=2\n",
      "\t7. 179,209,595 - 179,209,700, ID=ENSE00001077691, phase=2, end phase=0\n",
      "\t8. 179,210,186 - 179,210,338, ID=ENSE00001128470, phase=0, end phase=0\n",
      "\t9. 179,210,431 - 179,210,565, ID=ENSE00001128465, phase=0, end phase=0\n",
      "\t10. 179,218,210 - 179,218,334, ID=ENSE00001077674, phase=0, end phase=2\n",
      "\t11. 179,219,196 - 179,219,277, ID=ENSE00000826291, phase=2, end phase=0\n",
      "\t12. 179,219,571 - 179,219,735, ID=ENSE00000826292, phase=0, end phase=0\n",
      "\t13. 179,219,949 - 179,220,052, ID=ENSE00003485038, phase=0, end phase=2\n",
      "\t14. 179,220,986 - 179,221,157, ID=ENSE00003568097, phase=2, end phase=0\n",
      "\t15. 179,224,081 - 179,224,187, ID=ENSE00003489671, phase=0, end phase=2\n",
      "\t16. 179,224,700 - 179,224,821, ID=ENSE00003485539, phase=2, end phase=1\n",
      "\t17. 179,225,962 - 179,226,040, ID=ENSE00000826297, phase=1, end phase=2\n",
      "\t18. 179,229,272 - 179,229,442, ID=ENSE00000826298, phase=2, end phase=2\n",
      "\t19. 179,230,004 - 179,230,121, ID=ENSE00000826299, phase=2, end phase=0\n",
      "\t20. 179,230,225 - 179,230,376, ID=ENSE00000826300, phase=0, end phase=2\n",
      "\t21. 179,234,094 - 179,240,093, ID=ENSE00001139987, phase=2, end phase=-1\n",
      "CDS regions:\n",
      "\t1. 179,198,826 - 179,199,177, phase=0\n",
      "\t2. 179,199,690 - 179,199,899, phase=2\n",
      "\t3. 179,201,290 - 179,201,540, phase=2\n",
      "\t4. 179,203,544 - 179,203,789, phase=0\n",
      "\t5. 179,204,503 - 179,204,588, phase=0\n",
      "\t6. 179,209,595 - 179,209,700, phase=1\n",
      "\t7. 179,210,186 - 179,210,338, phase=0\n",
      "\t8. 179,210,431 - 179,210,565, phase=0\n",
      "\t9. 179,218,210 - 179,218,334, phase=0\n",
      "\t10. 179,219,196 - 179,219,277, phase=1\n",
      "\t11. 179,219,571 - 179,219,735, phase=0\n",
      "\t12. 179,219,949 - 179,220,052, phase=0\n",
      "\t13. 179,220,986 - 179,221,157, phase=1\n",
      "\t14. 179,224,081 - 179,224,187, phase=0\n",
      "\t15. 179,224,700 - 179,224,821, phase=1\n",
      "\t16. 179,225,962 - 179,226,040, phase=2\n",
      "\t17. 179,229,272 - 179,229,442, phase=1\n",
      "\t18. 179,230,004 - 179,230,121, phase=1\n",
      "\t19. 179,230,225 - 179,230,376, phase=0\n",
      "\t20. 179,234,094 - 179,234,364, phase=1\n",
      "5UTR regions:\n",
      "\t1. 179,148,357 - 179,148,603\n",
      "\t2. 179,198,750 - 179,198,825\n",
      "3UTR regions:\n",
      "\t1. 179,234,365 - 179,240,093\n"
     ]
    }
   ],
   "source": [
    "# get a detailed information about a transcript\n",
    "print(f\"Detailed {transcript_id=} information:\")\n",
    "gene_obj.transcript_info(transcript_id, verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Regarding the exon phases see [this](https://blog.gene-test.com/coding-phases-frames/). Basically:\n",
    "1. exon end phase is the exon phase of the next exon.\n",
    "2. Conversion from exon phase to CDS phase: 0-->0, 1-->2, 2-->1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Exon-intron map for transcript_id='ENST00000263967':\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>region</th>\n",
       "      <th>region_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Exon 1</td>\n",
       "      <td>(179148357, 179148603)</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Intron 1</td>\n",
       "      <td>(179148604, 179198749)</td>\n",
       "      <td>50146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Exon 2</td>\n",
       "      <td>(179198750, 179199177)</td>\n",
       "      <td>428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Intron 2</td>\n",
       "      <td>(179199178, 179199689)</td>\n",
       "      <td>512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Exon 3</td>\n",
       "      <td>(179199690, 179199899)</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Intron 3</td>\n",
       "      <td>(179199900, 179201289)</td>\n",
       "      <td>1390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Exon 4</td>\n",
       "      <td>(179201290, 179201540)</td>\n",
       "      <td>251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Intron 4</td>\n",
       "      <td>(179201541, 179203543)</td>\n",
       "      <td>2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Exon 5</td>\n",
       "      <td>(179203544, 179203789)</td>\n",
       "      <td>246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Intron 5</td>\n",
       "      <td>(179203790, 179204502)</td>\n",
       "      <td>713</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Exon 6</td>\n",
       "      <td>(179204503, 179204588)</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Intron 6</td>\n",
       "      <td>(179204589, 179209594)</td>\n",
       "      <td>5006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Exon 7</td>\n",
       "      <td>(179209595, 179209700)</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Intron 7</td>\n",
       "      <td>(179209701, 179210185)</td>\n",
       "      <td>485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Exon 8</td>\n",
       "      <td>(179210186, 179210338)</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Intron 8</td>\n",
       "      <td>(179210339, 179210430)</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Exon 9</td>\n",
       "      <td>(179210431, 179210565)</td>\n",
       "      <td>135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Intron 9</td>\n",
       "      <td>(179210566, 179218209)</td>\n",
       "      <td>7644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Exon 10</td>\n",
       "      <td>(179218210, 179218334)</td>\n",
       "      <td>125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Intron 10</td>\n",
       "      <td>(179218335, 179219195)</td>\n",
       "      <td>861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Exon 11</td>\n",
       "      <td>(179219196, 179219277)</td>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Intron 11</td>\n",
       "      <td>(179219278, 179219570)</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Exon 12</td>\n",
       "      <td>(179219571, 179219735)</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Intron 12</td>\n",
       "      <td>(179219736, 179219948)</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Exon 13</td>\n",
       "      <td>(179219949, 179220052)</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Intron 13</td>\n",
       "      <td>(179220053, 179220985)</td>\n",
       "      <td>933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Exon 14</td>\n",
       "      <td>(179220986, 179221157)</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Intron 14</td>\n",
       "      <td>(179221158, 179224080)</td>\n",
       "      <td>2923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Exon 15</td>\n",
       "      <td>(179224081, 179224187)</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Intron 15</td>\n",
       "      <td>(179224188, 179224699)</td>\n",
       "      <td>512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Exon 16</td>\n",
       "      <td>(179224700, 179224821)</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Intron 16</td>\n",
       "      <td>(179224822, 179225961)</td>\n",
       "      <td>1140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Exon 17</td>\n",
       "      <td>(179225962, 179226040)</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Intron 17</td>\n",
       "      <td>(179226041, 179229271)</td>\n",
       "      <td>3231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Exon 18</td>\n",
       "      <td>(179229272, 179229442)</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Intron 18</td>\n",
       "      <td>(179229443, 179230003)</td>\n",
       "      <td>561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Exon 19</td>\n",
       "      <td>(179230004, 179230121)</td>\n",
       "      <td>118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Intron 19</td>\n",
       "      <td>(179230122, 179230224)</td>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Exon 20</td>\n",
       "      <td>(179230225, 179230376)</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Intron 20</td>\n",
       "      <td>(179230377, 179234093)</td>\n",
       "      <td>3717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Exon 21</td>\n",
       "      <td>(179234094, 179240093)</td>\n",
       "      <td>6000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         name                  region  region_size\n",
       "0      Exon 1  (179148357, 179148603)          247\n",
       "1    Intron 1  (179148604, 179198749)        50146\n",
       "2      Exon 2  (179198750, 179199177)          428\n",
       "3    Intron 2  (179199178, 179199689)          512\n",
       "4      Exon 3  (179199690, 179199899)          210\n",
       "5    Intron 3  (179199900, 179201289)         1390\n",
       "6      Exon 4  (179201290, 179201540)          251\n",
       "7    Intron 4  (179201541, 179203543)         2003\n",
       "8      Exon 5  (179203544, 179203789)          246\n",
       "9    Intron 5  (179203790, 179204502)          713\n",
       "10     Exon 6  (179204503, 179204588)           86\n",
       "11   Intron 6  (179204589, 179209594)         5006\n",
       "12     Exon 7  (179209595, 179209700)          106\n",
       "13   Intron 7  (179209701, 179210185)          485\n",
       "14     Exon 8  (179210186, 179210338)          153\n",
       "15   Intron 8  (179210339, 179210430)           92\n",
       "16     Exon 9  (179210431, 179210565)          135\n",
       "17   Intron 9  (179210566, 179218209)         7644\n",
       "18    Exon 10  (179218210, 179218334)          125\n",
       "19  Intron 10  (179218335, 179219195)          861\n",
       "20    Exon 11  (179219196, 179219277)           82\n",
       "21  Intron 11  (179219278, 179219570)          293\n",
       "22    Exon 12  (179219571, 179219735)          165\n",
       "23  Intron 12  (179219736, 179219948)          213\n",
       "24    Exon 13  (179219949, 179220052)          104\n",
       "25  Intron 13  (179220053, 179220985)          933\n",
       "26    Exon 14  (179220986, 179221157)          172\n",
       "27  Intron 14  (179221158, 179224080)         2923\n",
       "28    Exon 15  (179224081, 179224187)          107\n",
       "29  Intron 15  (179224188, 179224699)          512\n",
       "30    Exon 16  (179224700, 179224821)          122\n",
       "31  Intron 16  (179224822, 179225961)         1140\n",
       "32    Exon 17  (179225962, 179226040)           79\n",
       "33  Intron 17  (179226041, 179229271)         3231\n",
       "34    Exon 18  (179229272, 179229442)          171\n",
       "35  Intron 18  (179229443, 179230003)          561\n",
       "36    Exon 19  (179230004, 179230121)          118\n",
       "37  Intron 19  (179230122, 179230224)          103\n",
       "38    Exon 20  (179230225, 179230376)          152\n",
       "39  Intron 20  (179230377, 179234093)         3717\n",
       "40    Exon 21  (179234094, 179240093)         6000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# exon-intron table\n",
    "print(f'\\nExon-intron map for {transcript_id=}:')\n",
    "display(gene_obj.exon_intron_map(transcript_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Exon map for transcript_id='ENST00000263967':\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_region</th>\n",
       "      <th>exon_size</th>\n",
       "      <th>mRNA_NT_region</th>\n",
       "      <th>ORF_NT_region</th>\n",
       "      <th>ORF_AA_region</th>\n",
       "      <th>next_exon_frame_alignment</th>\n",
       "      <th>exon_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>(179148357, 179148603)</td>\n",
       "      <td>247</td>\n",
       "      <td>[1, 247]</td>\n",
       "      <td>[0, 0]</td>\n",
       "      <td>[0, 0]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001493081</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>(179198750, 179199177)</td>\n",
       "      <td>428</td>\n",
       "      <td>[248, 675]</td>\n",
       "      <td>[1, 352]</td>\n",
       "      <td>[1, 118]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00001139995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>(179199690, 179199899)</td>\n",
       "      <td>210</td>\n",
       "      <td>[676, 885]</td>\n",
       "      <td>[353, 562]</td>\n",
       "      <td>[118, 188]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000997375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>(179201290, 179201540)</td>\n",
       "      <td>251</td>\n",
       "      <td>[886, 1136]</td>\n",
       "      <td>[563, 813]</td>\n",
       "      <td>[188, 271]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001077693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>(179203544, 179203789)</td>\n",
       "      <td>246</td>\n",
       "      <td>[1137, 1382]</td>\n",
       "      <td>[814, 1059]</td>\n",
       "      <td>[272, 353]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001077692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>(179204503, 179204588)</td>\n",
       "      <td>86</td>\n",
       "      <td>[1383, 1468]</td>\n",
       "      <td>[1060, 1145]</td>\n",
       "      <td>[354, 382]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001077694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>(179209595, 179209700)</td>\n",
       "      <td>106</td>\n",
       "      <td>[1469, 1574]</td>\n",
       "      <td>[1146, 1251]</td>\n",
       "      <td>[382, 417]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001077691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>(179210186, 179210338)</td>\n",
       "      <td>153</td>\n",
       "      <td>[1575, 1727]</td>\n",
       "      <td>[1252, 1404]</td>\n",
       "      <td>[418, 468]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001128470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>(179210431, 179210565)</td>\n",
       "      <td>135</td>\n",
       "      <td>[1728, 1862]</td>\n",
       "      <td>[1405, 1539]</td>\n",
       "      <td>[469, 513]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001128465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>(179218210, 179218334)</td>\n",
       "      <td>125</td>\n",
       "      <td>[1863, 1987]</td>\n",
       "      <td>[1540, 1664]</td>\n",
       "      <td>[514, 555]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001077674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>(179219196, 179219277)</td>\n",
       "      <td>82</td>\n",
       "      <td>[1988, 2069]</td>\n",
       "      <td>[1665, 1746]</td>\n",
       "      <td>[555, 582]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00000826291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>(179219571, 179219735)</td>\n",
       "      <td>165</td>\n",
       "      <td>[2070, 2234]</td>\n",
       "      <td>[1747, 1911]</td>\n",
       "      <td>[583, 637]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00000826292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>(179219949, 179220052)</td>\n",
       "      <td>104</td>\n",
       "      <td>[2235, 2338]</td>\n",
       "      <td>[1912, 2015]</td>\n",
       "      <td>[638, 672]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003485038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>(179220986, 179221157)</td>\n",
       "      <td>172</td>\n",
       "      <td>[2339, 2510]</td>\n",
       "      <td>[2016, 2187]</td>\n",
       "      <td>[672, 729]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00003568097</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>(179224081, 179224187)</td>\n",
       "      <td>107</td>\n",
       "      <td>[2511, 2617]</td>\n",
       "      <td>[2188, 2294]</td>\n",
       "      <td>[730, 765]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003489671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>(179224700, 179224821)</td>\n",
       "      <td>122</td>\n",
       "      <td>[2618, 2739]</td>\n",
       "      <td>[2295, 2416]</td>\n",
       "      <td>[765, 806]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003485539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>(179225962, 179226040)</td>\n",
       "      <td>79</td>\n",
       "      <td>[2740, 2818]</td>\n",
       "      <td>[2417, 2495]</td>\n",
       "      <td>[806, 832]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00000826297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>(179229272, 179229442)</td>\n",
       "      <td>171</td>\n",
       "      <td>[2819, 2989]</td>\n",
       "      <td>[2496, 2666]</td>\n",
       "      <td>[832, 889]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00000826298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>(179230004, 179230121)</td>\n",
       "      <td>118</td>\n",
       "      <td>[2990, 3107]</td>\n",
       "      <td>[2667, 2784]</td>\n",
       "      <td>[889, 928]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00000826299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>(179230225, 179230376)</td>\n",
       "      <td>152</td>\n",
       "      <td>[3108, 3259]</td>\n",
       "      <td>[2785, 2936]</td>\n",
       "      <td>[929, 979]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00000826300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>(179234094, 179240093)</td>\n",
       "      <td>6000</td>\n",
       "      <td>[3260, 9259]</td>\n",
       "      <td>[2937, 3207]</td>\n",
       "      <td>[979, 1068]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001139987</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    exon_number             exon_region  exon_size mRNA_NT_region  \\\n",
       "0             1  (179148357, 179148603)        247       [1, 247]   \n",
       "1             2  (179198750, 179199177)        428     [248, 675]   \n",
       "2             3  (179199690, 179199899)        210     [676, 885]   \n",
       "3             4  (179201290, 179201540)        251    [886, 1136]   \n",
       "4             5  (179203544, 179203789)        246   [1137, 1382]   \n",
       "5             6  (179204503, 179204588)         86   [1383, 1468]   \n",
       "6             7  (179209595, 179209700)        106   [1469, 1574]   \n",
       "7             8  (179210186, 179210338)        153   [1575, 1727]   \n",
       "8             9  (179210431, 179210565)        135   [1728, 1862]   \n",
       "9            10  (179218210, 179218334)        125   [1863, 1987]   \n",
       "10           11  (179219196, 179219277)         82   [1988, 2069]   \n",
       "11           12  (179219571, 179219735)        165   [2070, 2234]   \n",
       "12           13  (179219949, 179220052)        104   [2235, 2338]   \n",
       "13           14  (179220986, 179221157)        172   [2339, 2510]   \n",
       "14           15  (179224081, 179224187)        107   [2511, 2617]   \n",
       "15           16  (179224700, 179224821)        122   [2618, 2739]   \n",
       "16           17  (179225962, 179226040)         79   [2740, 2818]   \n",
       "17           18  (179229272, 179229442)        171   [2819, 2989]   \n",
       "18           19  (179230004, 179230121)        118   [2990, 3107]   \n",
       "19           20  (179230225, 179230376)        152   [3108, 3259]   \n",
       "20           21  (179234094, 179240093)       6000   [3260, 9259]   \n",
       "\n",
       "   ORF_NT_region ORF_AA_region  next_exon_frame_alignment          exon_ID  \n",
       "0         [0, 0]        [0, 0]                          0  ENSE00001493081  \n",
       "1       [1, 352]      [1, 118]                          2  ENSE00001139995  \n",
       "2     [353, 562]    [118, 188]                          2  ENSE00000997375  \n",
       "3     [563, 813]    [188, 271]                          0  ENSE00001077693  \n",
       "4    [814, 1059]    [272, 353]                          0  ENSE00001077692  \n",
       "5   [1060, 1145]    [354, 382]                          1  ENSE00001077694  \n",
       "6   [1146, 1251]    [382, 417]                          0  ENSE00001077691  \n",
       "7   [1252, 1404]    [418, 468]                          0  ENSE00001128470  \n",
       "8   [1405, 1539]    [469, 513]                          0  ENSE00001128465  \n",
       "9   [1540, 1664]    [514, 555]                          1  ENSE00001077674  \n",
       "10  [1665, 1746]    [555, 582]                          0  ENSE00000826291  \n",
       "11  [1747, 1911]    [583, 637]                          0  ENSE00000826292  \n",
       "12  [1912, 2015]    [638, 672]                          1  ENSE00003485038  \n",
       "13  [2016, 2187]    [672, 729]                          0  ENSE00003568097  \n",
       "14  [2188, 2294]    [730, 765]                          1  ENSE00003489671  \n",
       "15  [2295, 2416]    [765, 806]                          2  ENSE00003485539  \n",
       "16  [2417, 2495]    [806, 832]                          1  ENSE00000826297  \n",
       "17  [2496, 2666]    [832, 889]                          1  ENSE00000826298  \n",
       "18  [2667, 2784]    [889, 928]                          0  ENSE00000826299  \n",
       "19  [2785, 2936]    [929, 979]                          1  ENSE00000826300  \n",
       "20  [2937, 3207]   [979, 1068]                          0  ENSE00001139987  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# exon table\n",
    "print(f'\\nExon map for {transcript_id=}:')\n",
    "display(gene_obj.exon_map(transcript_id))\n",
    "\n",
    "# if (df := gene_obj.exon_map(transcript_id)) is not None:\n",
    "#     print(f\"Exon map for {gene=}, {transcript_id=}:\")\n",
    "#     display(df.style.format(formatter={\"exon_region\": lambda x: f'({x[0]:,}, {x[1]:,})'}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "exon map can be save to excel or html files.\n",
    "'''\n",
    "gene_obj.exon_map_to_html(transcript_id, report_path / f\"{gene}_{transcript_id}_exon_map.html\")\n",
    "gene_obj.exon_map_to_excel(transcript_id, report_path / f\"{gene}_{transcript_id}_exon_map.xlsx\", usr_desc={\"General\": f\"Exon table for {gene}, {transcript_id}\", \"Source\": \"gene_annotation_example.ipynb\" })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "transcript_id='ENST00000263967' contains 91,737 bps.\n",
      "\n",
      "rna=\n",
      "AGTTCCGGTGCCGCCGCTGCGGCCGCTGAGGTGTCGGGCTGCTGCTGCCGCGGCCGCTGGGACTGGGGCTGGGGCCGCCGGCGAGGCAGGGCTCGGGCCCGGCCGGGCAGCTCCGGAGCGGCGGGGGAGAGGGGCCGGGAGGCGGGGGCCGTGCCGCCCGCTCTCCTCTCCCTCGGCGCCGCCGCCGCCGCCCGCGGGGCTGGGACCCGATGCGGTTAGAGCCGCGGAGCCTGGAAGAGCCCCGAGCGTTTCTGCTTTGGGACAACCATACATCTAATTCCTTAAAGTAGTTTTATATGTAAAACTTGCAAAGAATCAGAACAATGCCTCCACGACCATCATCAGGTGAACTGTGGGGCATCCACTTGATGCCCCCAAGAATCCTAGTAGAATGTTTACTACCAAATGGAATGATAGTGACTTTAGAATGCCTCCGTGAGGCTACATTAATAACCATAAAGCATGAACTATTTAAAGAAGCAAGAAAATACCCCCTCCATCAACTTCTTCAAGATGAATCTTCTTACATTTTCGTAAGTGTTACTCAAGAAGCAGAAAGGGAAGAATTTTTTGATGAAACAAGACGACTTTGTGACCTTCGGCTTTTTCAACCCTTTTTAAAAGTAATTGAACCAGTAGGCAACCGTGAAGAAAAGATCCTCAATCGAGAAATTGGTTTTGCTATCGGCATGCCAGTGTGTGAATTTGATATGGTTAAAGATCCAGAAGTACAGGACTTCCGAAGAAATATTCTGAACGTTTGTAAAGAAGCTGTGGATCTTAGGGACCTCAATTCACCTCATAGTAGAGCAATGTATGTCTATCCTCCAAATGTAGAATCTTCACCAGAATTGCCAAAGCACATATATAATAAATTAGATAAAGGGCAAATAATAGTGGTGATCTGGGTAATAGTTTCTCCAAATAATGACAAGCAGAAGTATACTCTGAAAATCAACCATGACTGTGTACCAGAACAAGTAATTGCTGAAGCAATCAGGAAAAAAACTCGAAGTATGTTGCTATCCTCTGAACAACTAAAACTCTGTGTTTTAGAATATCAGGGCAAGTATATTTTAAAAGTGTGTGGATGTGATGAATACTTCCTAGAAAAATATCCTCTGAGTCAGTATAAGTATATAAGAAGCTGTATAATGCTTGGGAGGATGCCCAATTTGATGTTGATGGCTAAAGAAAGCCTTTATTCTCAACTGCCAATGGACTGTTTTACAATGCCATCTTATTCCAGACGCATTTCCACAGCTACACCATATATGAATGGAGAAACATCTACAAAATCCCTTTGGGTTATAAATAGTGCACTCAGAATAAAAATTCTTTGTGCAACCTACGTGAATGTAAATATTCGAGACATTGATAAGATCTATGTTCGAACAGGTATCTACCATGGAGGAGAACCCTTATGTGACAATGTGAACACTCAAAGAGTACCTTGTTCCAATCCCAGGTGGAATGAATGGCTGAATTATGATATATACATTCCTGATCTTCCTCGTGCTGCTCGACTTTGCCTTTCCATTTGCTCTGTTAAAGGCCGAAAGGGTGCTAAAGAGGAACACTGTCCATTGGCATGGGGAAATATAAACTTGTTTGATTACACAGACACTCTAGTATCTGGAAAAATGGCTTTGAATCTTTGGCCAGTACCTCATGGATTAGAAGATTTGCTGAACCCTATTGGTGTTACTGGATCAAATCCAAATAAAGAAACTCCATGCTTAGAGTTGGAGTTTGACTGGTTCAGCAGTGTGGTAAAGTTCCCAGATATGTCAGTGATTGAAGAGCATGCCAATTGGTCTGTATCCCGAGAAGCAGGATTTAGCTATTCCCACGCAGGACTGAGTAACAGACTAGCTAGAGACAATGAATTAAGGGAAAATGACAAAGAACAGCTCAAAGCAATTTCTACACGAGATCCTCTCTCTGAAATCACTGAGCAGGAGAAAGATTTTCTATGGAGTCACAGACACTATTGTGTAACTATCCCCGAAATTCTACCCAAATTGCTTCTGTCTGTTAAATGGAATTCTAGAGATGAAGTAGCCCAGATGTATTGCTTGGTAAAAGATTGGCCTCCAATCAAACCTGAACAGGCTATGGAACTTCTGGACTGTAATTACCCAGATCCTATGGTTCGAGGTTTTGCTGTTCGGTGCTTGGAAAAATATTTAACAGATGACAAACTTTCTCAGTATTTAATTCAGCTAGTACAGGTCCTAAAATATGAACAATATTTGGATAACTTGCTTGTGAGATTTTTACTGAAGAAAGCATTGACTAATCAAAGGATTGGGCACTTTTTCTTTTGGCATTTAAAATCTGAGATGCACAATAAAACAGTTAGCCAGAGGTTTGGCCTGCTTTTGGAGTCCTATTGTCGTGCATGTGGGATGTATTTGAAGCACCTGAATAGGCAAGTCGAGGCAATGGAAAAGCTCATTAACTTAACTGACATTCTCAAACAGGAGAAGAAGGATGAAACACAAAAGGTACAGATGAAGTTTTTAGTTGAGCAAATGAGGCGACCAGATTTCATGGATGCTCTACAGGGCTTTCTGTCTCCTCTAAACCCTGCTCATCAACTAGGAAACCTCAGGCTTGAAGAGTGTCGAATTATGTCCTCTGCAAAAAGGCCACTGTGGTTGAATTGGGAGAACCCAGACATCATGTCAGAGTTACTGTTTCAGAACAATGAGATCATCTTTAAAAATGGGGATGATTTACGGCAAGATATGCTAACACTTCAAATTATTCGTATTATGGAAAATATCTGGCAAAATCAAGGTCTTGATCTTCGAATGTTACCTTATGGTTGTCTGTCAATCGGTGACTGTGTGGGACTTATTGAGGTGGTGCGAAATTCTCACACTATTATGCAAATTCAGTGCAAAGGCGGCTTGAAAGGTGCACTGCAGTTCAACAGCCACACACTACATCAGTGGCTCAAAGACAAGAACAAAGGAGAAATATATGATGCAGCCATTGACCTGTTTACACGTTCATGTGCTGGATACTGTGTAGCTACCTTCATTTTGGGAATTGGAGATCGTCACAATAGTAACATCATGGTGAAAGACGATGGACAACTGTTTCATATAGATTTTGGACACTTTTTGGATCACAAGAAGAAAAAATTTGGTTATAAACGAGAACGTGTGCCATTTGTTTTGACACAGGATTTCTTAATAGTGATTAGTAAAGGAGCCCAAGAATGCACAAAGACAAGAGAATTTGAGAGGTTTCAGGAGATGTGTTACAAGGCTTATCTAGCTATTCGACAGCATGCCAATCTCTTCATAAATCTTTTCTCAATGATGCTTGGCTCTGGAATGCCAGAACTACAATCTTTTGATGACATTGCATACATTCGAAAGACCCTAGCCTTAGATAAAACTGAGCAAGAGGCTTTGGAGTATTTCATGAAACAAATGAATGATGCACATCATGGTGGCTGGACAACAAAAATGGATTGGATCTTCCACACAATTAAACAGCATGCATTGAACTGAAAAGATAACTGAGAAAATGAAAGCTCACTCTGGATTCCACACTGCACTGTTAATAACTCTCAGCAGGCAAAGACCGATTGCATAGGAATTGCACAATCCATGAACAGCATTAGAATTTACAGCAAGAACAGAAATAAAATACTATATAATTTAAATAATGTAAACGCAAACAGGGTTTGATAGCACTTAAACTAGTTCATTTCAAAATTAAGCTTTAGAATAATGCGCAATTTCATGTTATGCCTTAAGTCCAAAAAGGTAAACTTTGAAGATTGTTTGTATCTTTTTTTAAAAAACAAAACAAAACAAAAATCCCCAAAATATATAGAAATGATGGAGAAGGAAAAAGTGATGGTTTTTTTTGTCTTGCAAATGTTCTATGTTTTGAAATGTGGACACAACAAAGGCTGTTATTGCATTAGGTGTAAGTAAACTGGAGTTTATGTTAAATTACATTGATTGGAAAAGAATGAAAATTTCTTATTTTTCCATTGCTGTTCAATTTATAGTTTGAAGTGGGTTTTTGACTGCTTGTTTAATGAAGAAAAATGCTTGGGGTGGAAGGGACTCTTGAGATTTCACCAGAGACTTTTTCTTTTTAATAAATCAAACCTTTTGATGATTTGAGGTTTTATCTGCAGTTTTGGAAGCAGTCACAAATGAGACCTGTTATAAGGTGGTATTTTTTTTTTTCTTCTGGACAGTATTTAAAGGATCTTATTCTTATTTCCCAGGGAAATTCTGGGCTCCCACAAAGTAAAAAAAAAAAAAAATCATAGAAAAAGAATGAGCAGGAATAGTTCTTATTCCAGAATTGTACAGTATTCACCTTAAGTTGATTTTTTTTCTCCTTCTGCAATTGAACTGAATACATTTTTCATGCATGTTTTCCAGAAAATAGAAGTATTAATGTTATTAAAAAGATTATTTTTTTTATTAAAGGCTATTTATATTATAGAAACTATCATTAATATATATTCTTTATTTACATGATCTGTCCCATAGTCATGCATTGTTTTGCACCCCAAATTTTTTATTGTTCATAGCAGCATGGTCAGCTTTCTTCTTGATCTATAGATGAGGCTCAGGCACTATCCCATTTATACCAATAACCAGTGTATAACTACTTAAGGAAAACATAAAAACTTCATCTTCTTTCCTTTTATTTCTTATGTGAATCTCCCGTCTTCCATTCTCTTTTATAATTGAGAATGTCTCAATCATATGAAATTAGTTACCAGAATTAACACAATTTAGACTATCTTCCTGATTCCTTAAACCCCTTTACTGAAGTATACTCATGAATAATACTTTAAAATATGGGGGAATAGAAACCATGAACTTTTTACCTTTTTAAACTATTTATCCATATCTCCAAAGTAGAACATTAAACCATTTTAAGATATGTCTCATTCCCAAGTAGTCAGAGCTCACTCTCCAACTTTATTAAATACTATTTGAGCACAGGACACATTCTTAAACATTTTGAAAAACATTAACCCAAGATGTAGAGGCTACTGCTAGTCGTCATTCTAGAATCTGATATTTTACTCTGTATTTGAAATGAATGATTAATGTCCTAGGAAATTAGCTTTAGCAGATGTCCAGGTGCCACATCAAAAAAGTGCAATAATTATTGACAGTTTTTTAGATTAGGCATATTATTGGAAAACAACTTTATAAAGAGTGAACATTGTATACTCTAGTAAAACAGCATCACTTTAAAAATATTCATTTATGAAATCTGTTACCTATAGTTGAAGTCTTGAGTAGTGAACAAGGGACTCTAATACCAATACTCTTAATATCTGGCTATTTTAGATCCCTTAAAGGGCATAATTATTGGAAATTTAGGTATTTCACTAAAGCATGTATATAATATTGCCAACAAGAAAAGTAAATTTGAAGATTAAGGGAACTTACTTCTGCAAACTGTCTTGCGATAGTTAAGCAGAATTTAAACTCTGTTTTAAGCAGGAAACCAGAAAGATTATTTTGCAGTTGTAGAAGATTTCATAACTTATTAAAACTTATTAACATTTTGTGTTGTTTAGATATAGGCAGTTGATACATACTAACATCCCAGCCTTTTCAATATCAGGGTTAAATTATAGGAAAACTCAGTAAAATGGTACAAATCTGAAAGTTTGATGGTAGAAACTGAAGATTTAACAGAGAACTGTGTTTTACCCGAGTGCCAAAAATGCTGTGAGCCTCCTTGCACAAAATTTATACCACTTTTGCATTTTTATCTATCAGTCCAGATAGTTGTCTCCCCTCCTTCTCCCAGGACCTCTCCACCATTAAAATGCACAAACCACATGGCCGATTTCACCATTTACATTTATTTTCAAAAGTTACTACAACCAAATTAATTCTATTAGAAGAAATGTAGACAAATTCTATAAAGACTATAGATTGTGACCTAAGAAAGAAATGAGGCAAAGAACCAAACATTGAATTAAATGCTACATGGGTGACTAAGATCTGTTTCAAGTCAGTGATAATATAGCCACTTCTGGGTACTTCAGTATCAGAGATCAGTTCTCGTGGTTTAGACAGTTCCTATCTATAGCTGACTATCCTTGTCCTTGAATATGGTGTAACTGACTATTGGCTCTACAGTTTTATTGGGCCACTTAAGAAATATTTCCTTGAATAATTATTTTGAGAAAAAGTCTAAAAGTAATAAAAATAATTTTAAACACACTGTAGTAAGAAATGACTGTTGGAAAATTATGCTTTCACTTTCTACCATATTCTCAGCTATACAAAACCATTTATTTTGAAGATTTTTAGACTACTGTTAATTTGAAATCTGTTACTCTTATTGTGGAATTTGTTTTTTTAAAAAAGATGTTTCTAATTGGATTTTTAAAAGAAGAATGGAATTTGGTTGCTATTTTACAATAGAACCTAAGCTTTTTGTGGTTCTTAGTGTCCTATGTAAAACTTAGTGTCAAAGTAATCAACTTTGAGATTTTCCCTTCTATTCTGCTTTATATTAAAAGCCCATTAGAAAATGGGAACCTGGTGAATATATAATGAATTGTAAAATATTTTAATGTGTAACTTTTTCAACTGTGAAACTGACTTGATTTTTTGATGAAAACAGCTGCTGATAAAGTATTTTGTGTAAAGTGTAGTTCTTATTAATCAGGAAAATGATGACTTGATTAGACTGTATATGCCCTCTTGGATTTTATTTTAAATGGATTGGTGACTTTCACATAGGTAAAACACAGTCCATCTGTATTCTTTTTTCCATCAAAAATCGAGTGATTTGGAATTATAAAAAAATTGTGAGCAGCCTATTTGAAAGGCATCATGGAAATTTCACAGCACAATAACACGGATTTGTTTTTTCTTAATGATGTAAATCCGTTTAATTCATACTTTGATCAATAGCCCATGCTTGCCAACTCTGAAGAAATTTAATTTCCAGCAGTATTTTAAAGCTAGCCTGTTAACTTTTTCTGAATATTTAAAGTTCCTCTTTTTTCTATGTCTGCACAAACTGCAGACCTGGGCTGGACCCACATACTCAAGAGTCCACCTTAAGAAATTATTTTGATGTCCAAGACATCACTAAAATATTTAAGTTTAAAGATAATATGTGGTGTTAATAGATTGTGGTGCTTTTACTATTTAAAGACAACTTTCATACTTCAGATGTTTTTGAGAAGAGGGGAATGTGAGGGGAGGGGGCAGAACAGGGAGGAGTTGTTTGAATGAATTACATTCTTTATATCCATCCTGCTCATTTGGGGCATGTCTTTAAGAGAAGGCTGAAAGTTGTGAGAGTATATTGTATACCGTAAGAGAATCAACTCTTCATCATGGATGGGATTGTGAAGGCTGAACTATAAAATTCAGCATTGACAGCATCCTCAATTAATAATTCTTGGTGACAGAATAATACAGCTGGGCTGTTTTTTAAAATATAAACAATACCATTTTTAATTATTACATTAAAAATTGTAAATATATCTATGTGCCATGGCCTGGGAAGCCTGCTTTCTTTTTTCATAAAAATTATTTTTACTGTATGAAAAGATCATGGGGTTTAGCTCAAAATATCTGTGGTCCTGATAAAATTGGATTGGTAACTCTACCTCAGAAGGAAAATGGGAAAAAAAAATAGATGAGTCACAATTCAATACTTCAAGCTCAGAAACTGTGCAGATCACTGAATTTTAGATTTATAAAGTCAGAGTTGGCATGCCTTGTTTTTAATGATATGGAAGACCTTAAGAAAAAAACTTGGCTGAAGTTTAATCGTTGGTCCAGCCATTTGAAAAAGGCAATAGTTTGAGGAGGTTCCCGAATTCGGCATTTGAAATTCATTTTGTTCTCTCTTCTTCATTATTAGTGCATTTGGTGTGTGTATACTTGCACACAATTCTGTTTGTGTACACACTGCTTGCTTAGCCCTAGTCAAGAGGCATCTTTTATAAAAGGTGTAAAGAAATATCAAGGTTCTAAAATTCGGAAGAGTTTAGAATTTATTAGGAGTTTCCCAAGTTGGGATGTTAGTCTTTAAATAAACTTCATGCACCTATTCCACTTAAGGTTTTGCACCTCCTTTTTATTAGTGCAGTGCCATTTCTTCTGCTTGATTTTAGGTATGTTAATATTCCAGCCTTGCTAGTTAGCATAAAGTGACAGGTGTGAGCCATGAGGAAATTTTCTGACTTAATTTGTACACAACTACATATAAGAGTTTTAGTGGAGGAAAAAAATTAGTCCCTTGTGCGTATACAGTAGTTAGGTAAATGATTTTTCTACCAACAGTATACTCCATTCCTCATGTAGGTAAGTACAGAAAAGGTTTTTAAATGTATTTTTTTAGCCAGTTAAAGTCTATGAATCTATCTGCAACCTTATTTAATCTGTCACTATAATAATTTTGTGGTTATGCTAAGAACCATGTATACTTTTAGGTATTCTTATTTTTGTCAATTTTTCTAGGTTGGCAAGGAGGCAGAAAACCTTCATTGTTTCATATTAAAATATAATTAGACTAAACTTAATTCTAGTATGAATTTCCAAAATCATTATCTATTTATTTCATTTTTATTTAATTTTGTTTTTATTTCATTTTTAAAAGTCCCTTGTTCAATTTAACTTATGTTCCTAAGAGAGGTTGGAGAACTTGGCCTTCATCTGATTTCAAAAATGTTTTGAGTTTCAAATGAAGTTAATGGTTTCAGTGTGATTCAGTCCTCAGACCTAATTGGGTTGAATAAAATCTAAAAGAATATACCCTTTTGGAGCATAACATTTTAATACCTTGGGGAATGTGGCACTACCAAAAGAAGACTACTAACACGTCAGATGTTCACCTGGAAGCTTTATCAAGAAATTCGAACCACCCTTTTGGCCCCATTAATTGTAGCAAGTTTATTTCTCTATATTTTGTCATTCAGTGAATTGAAGTCCTGTGGTATACTGCATTCATTAGAAGAAAAACGTTTTTAATGTCCTTTTAATGATGGCCCAGAAAGCATTTGACACAGCAAGATGCATGTGTTACTATATTGAGAATATAGAATAATAACAGTATCACTAAATTTAAGACCTCTTCCCAGTCTTGCTGTTCCTAGCAAGAAGTTTGGCCTGTGACTGCACTTACTGTTTATGCTCATCAGAAACTGTCAATGTCTGCTTTTCTTTAACTCTGCAGTCTGTAACATCACGCTGTTTATTAAAAAAAAAAAGAAAAATTA\n",
      "\n",
      "orf=\n",
      "ATGCCTCCACGACCATCATCAGGTGAACTGTGGGGCATCCACTTGATGCCCCCAAGAATCCTAGTAGAATGTTTACTACCAAATGGAATGATAGTGACTTTAGAATGCCTCCGTGAGGCTACATTAATAACCATAAAGCATGAACTATTTAAAGAAGCAAGAAAATACCCCCTCCATCAACTTCTTCAAGATGAATCTTCTTACATTTTCGTAAGTGTTACTCAAGAAGCAGAAAGGGAAGAATTTTTTGATGAAACAAGACGACTTTGTGACCTTCGGCTTTTTCAACCCTTTTTAAAAGTAATTGAACCAGTAGGCAACCGTGAAGAAAAGATCCTCAATCGAGAAATTGGTTTTGCTATCGGCATGCCAGTGTGTGAATTTGATATGGTTAAAGATCCAGAAGTACAGGACTTCCGAAGAAATATTCTGAACGTTTGTAAAGAAGCTGTGGATCTTAGGGACCTCAATTCACCTCATAGTAGAGCAATGTATGTCTATCCTCCAAATGTAGAATCTTCACCAGAATTGCCAAAGCACATATATAATAAATTAGATAAAGGGCAAATAATAGTGGTGATCTGGGTAATAGTTTCTCCAAATAATGACAAGCAGAAGTATACTCTGAAAATCAACCATGACTGTGTACCAGAACAAGTAATTGCTGAAGCAATCAGGAAAAAAACTCGAAGTATGTTGCTATCCTCTGAACAACTAAAACTCTGTGTTTTAGAATATCAGGGCAAGTATATTTTAAAAGTGTGTGGATGTGATGAATACTTCCTAGAAAAATATCCTCTGAGTCAGTATAAGTATATAAGAAGCTGTATAATGCTTGGGAGGATGCCCAATTTGATGTTGATGGCTAAAGAAAGCCTTTATTCTCAACTGCCAATGGACTGTTTTACAATGCCATCTTATTCCAGACGCATTTCCACAGCTACACCATATATGAATGGAGAAACATCTACAAAATCCCTTTGGGTTATAAATAGTGCACTCAGAATAAAAATTCTTTGTGCAACCTACGTGAATGTAAATATTCGAGACATTGATAAGATCTATGTTCGAACAGGTATCTACCATGGAGGAGAACCCTTATGTGACAATGTGAACACTCAAAGAGTACCTTGTTCCAATCCCAGGTGGAATGAATGGCTGAATTATGATATATACATTCCTGATCTTCCTCGTGCTGCTCGACTTTGCCTTTCCATTTGCTCTGTTAAAGGCCGAAAGGGTGCTAAAGAGGAACACTGTCCATTGGCATGGGGAAATATAAACTTGTTTGATTACACAGACACTCTAGTATCTGGAAAAATGGCTTTGAATCTTTGGCCAGTACCTCATGGATTAGAAGATTTGCTGAACCCTATTGGTGTTACTGGATCAAATCCAAATAAAGAAACTCCATGCTTAGAGTTGGAGTTTGACTGGTTCAGCAGTGTGGTAAAGTTCCCAGATATGTCAGTGATTGAAGAGCATGCCAATTGGTCTGTATCCCGAGAAGCAGGATTTAGCTATTCCCACGCAGGACTGAGTAACAGACTAGCTAGAGACAATGAATTAAGGGAAAATGACAAAGAACAGCTCAAAGCAATTTCTACACGAGATCCTCTCTCTGAAATCACTGAGCAGGAGAAAGATTTTCTATGGAGTCACAGACACTATTGTGTAACTATCCCCGAAATTCTACCCAAATTGCTTCTGTCTGTTAAATGGAATTCTAGAGATGAAGTAGCCCAGATGTATTGCTTGGTAAAAGATTGGCCTCCAATCAAACCTGAACAGGCTATGGAACTTCTGGACTGTAATTACCCAGATCCTATGGTTCGAGGTTTTGCTGTTCGGTGCTTGGAAAAATATTTAACAGATGACAAACTTTCTCAGTATTTAATTCAGCTAGTACAGGTCCTAAAATATGAACAATATTTGGATAACTTGCTTGTGAGATTTTTACTGAAGAAAGCATTGACTAATCAAAGGATTGGGCACTTTTTCTTTTGGCATTTAAAATCTGAGATGCACAATAAAACAGTTAGCCAGAGGTTTGGCCTGCTTTTGGAGTCCTATTGTCGTGCATGTGGGATGTATTTGAAGCACCTGAATAGGCAAGTCGAGGCAATGGAAAAGCTCATTAACTTAACTGACATTCTCAAACAGGAGAAGAAGGATGAAACACAAAAGGTACAGATGAAGTTTTTAGTTGAGCAAATGAGGCGACCAGATTTCATGGATGCTCTACAGGGCTTTCTGTCTCCTCTAAACCCTGCTCATCAACTAGGAAACCTCAGGCTTGAAGAGTGTCGAATTATGTCCTCTGCAAAAAGGCCACTGTGGTTGAATTGGGAGAACCCAGACATCATGTCAGAGTTACTGTTTCAGAACAATGAGATCATCTTTAAAAATGGGGATGATTTACGGCAAGATATGCTAACACTTCAAATTATTCGTATTATGGAAAATATCTGGCAAAATCAAGGTCTTGATCTTCGAATGTTACCTTATGGTTGTCTGTCAATCGGTGACTGTGTGGGACTTATTGAGGTGGTGCGAAATTCTCACACTATTATGCAAATTCAGTGCAAAGGCGGCTTGAAAGGTGCACTGCAGTTCAACAGCCACACACTACATCAGTGGCTCAAAGACAAGAACAAAGGAGAAATATATGATGCAGCCATTGACCTGTTTACACGTTCATGTGCTGGATACTGTGTAGCTACCTTCATTTTGGGAATTGGAGATCGTCACAATAGTAACATCATGGTGAAAGACGATGGACAACTGTTTCATATAGATTTTGGACACTTTTTGGATCACAAGAAGAAAAAATTTGGTTATAAACGAGAACGTGTGCCATTTGTTTTGACACAGGATTTCTTAATAGTGATTAGTAAAGGAGCCCAAGAATGCACAAAGACAAGAGAATTTGAGAGGTTTCAGGAGATGTGTTACAAGGCTTATCTAGCTATTCGACAGCATGCCAATCTCTTCATAAATCTTTTCTCAATGATGCTTGGCTCTGGAATGCCAGAACTACAATCTTTTGATGACATTGCATACATTCGAAAGACCCTAGCCTTAGATAAAACTGAGCAAGAGGCTTTGGAGTATTTCATGAAACAAATGAATGATGCACATCATGGTGGCTGGACAACAAAAATGGATTGGATCTTCCACACAATTAAACAGCATGCATTGAACTGA\n",
      "\n",
      "AA=\n",
      "MPPRPSSGELWGIHLMPPRILVECLLPNGMIVTLECLREATLITIKHELFKEARKYPLHQLLQDESSYIFVSVTQEAEREEFFDETRRLCDLRLFQPFLKVIEPVGNREEKILNREIGFAIGMPVCEFDMVKDPEVQDFRRNILNVCKEAVDLRDLNSPHSRAMYVYPPNVESSPELPKHIYNKLDKGQIIVVIWVIVSPNNDKQKYTLKINHDCVPEQVIAEAIRKKTRSMLLSSEQLKLCVLEYQGKYILKVCGCDEYFLEKYPLSQYKYIRSCIMLGRMPNLMLMAKESLYSQLPMDCFTMPSYSRRISTATPYMNGETSTKSLWVINSALRIKILCATYVNVNIRDIDKIYVRTGIYHGGEPLCDNVNTQRVPCSNPRWNEWLNYDIYIPDLPRAARLCLSICSVKGRKGAKEEHCPLAWGNINLFDYTDTLVSGKMALNLWPVPHGLEDLLNPIGVTGSNPNKETPCLELEFDWFSSVVKFPDMSVIEEHANWSVSREAGFSYSHAGLSNRLARDNELRENDKEQLKAISTRDPLSEITEQEKDFLWSHRHYCVTIPEILPKLLLSVKWNSRDEVAQMYCLVKDWPPIKPEQAMELLDCNYPDPMVRGFAVRCLEKYLTDDKLSQYLIQLVQVLKYEQYLDNLLVRFLLKKALTNQRIGHFFFWHLKSEMHNKTVSQRFGLLLESYCRACGMYLKHLNRQVEAMEKLINLTDILKQEKKDETQKVQMKFLVEQMRRPDFMDALQGFLSPLNPAHQLGNLRLEECRIMSSAKRPLWLNWENPDIMSELLFQNNEIIFKNGDDLRQDMLTLQIIRIMENIWQNQGLDLRMLPYGCLSIGDCVGLIEVVRNSHTIMQIQCKGGLKGALQFNSHTLHQWLKDKNKGEIYDAAIDLFTRSCAGYCVATFILGIGDRHNSNIMVKDDGQLFHIDFGHFLDHKKKKFGYKRERVPFVLTQDFLIVISKGAQECTKTREFERFQEMCYKAYLAIRQHANLFINLFSMMLGSGMPELQSFDDIAYIRKTLALDKTEQEALEYFMKQMNDAHHGGWTTKMDWIFHTIKQHALN\n",
      "\n",
      "utr5=\n",
      "AGTTCCGGTGCCGCCGCTGCGGCCGCTGAGGTGTCGGGCTGCTGCTGCCGCGGCCGCTGGGACTGGGGCTGGGGCCGCCGGCGAGGCAGGGCTCGGGCCCGGCCGGGCAGCTCCGGAGCGGCGGGGGAGAGGGGCCGGGAGGCGGGGGCCGTGCCGCCCGCTCTCCTCTCCCTCGGCGCCGCCGCCGCCGCCCGCGGGGCTGGGACCCGATGCGGTTAGAGCCGCGGAGCCTGGAAGAGCCCCGAGCGTTTCTGCTTTGGGACAACCATACATCTAATTCCTTAAAGTAGTTTTATATGTAAAACTTGCAAAGAATCAGAACA\n",
      "\n",
      "utr3=\n",
      "AAAGATAACTGAGAAAATGAAAGCTCACTCTGGATTCCACACTGCACTGTTAATAACTCTCAGCAGGCAAAGACCGATTGCATAGGAATTGCACAATCCATGAACAGCATTAGAATTTACAGCAAGAACAGAAATAAAATACTATATAATTTAAATAATGTAAACGCAAACAGGGTTTGATAGCACTTAAACTAGTTCATTTCAAAATTAAGCTTTAGAATAATGCGCAATTTCATGTTATGCCTTAAGTCCAAAAAGGTAAACTTTGAAGATTGTTTGTATCTTTTTTTAAAAAACAAAACAAAACAAAAATCCCCAAAATATATAGAAATGATGGAGAAGGAAAAAGTGATGGTTTTTTTTGTCTTGCAAATGTTCTATGTTTTGAAATGTGGACACAACAAAGGCTGTTATTGCATTAGGTGTAAGTAAACTGGAGTTTATGTTAAATTACATTGATTGGAAAAGAATGAAAATTTCTTATTTTTCCATTGCTGTTCAATTTATAGTTTGAAGTGGGTTTTTGACTGCTTGTTTAATGAAGAAAAATGCTTGGGGTGGAAGGGACTCTTGAGATTTCACCAGAGACTTTTTCTTTTTAATAAATCAAACCTTTTGATGATTTGAGGTTTTATCTGCAGTTTTGGAAGCAGTCACAAATGAGACCTGTTATAAGGTGGTATTTTTTTTTTTCTTCTGGACAGTATTTAAAGGATCTTATTCTTATTTCCCAGGGAAATTCTGGGCTCCCACAAAGTAAAAAAAAAAAAAAATCATAGAAAAAGAATGAGCAGGAATAGTTCTTATTCCAGAATTGTACAGTATTCACCTTAAGTTGATTTTTTTTCTCCTTCTGCAATTGAACTGAATACATTTTTCATGCATGTTTTCCAGAAAATAGAAGTATTAATGTTATTAAAAAGATTATTTTTTTTATTAAAGGCTATTTATATTATAGAAACTATCATTAATATATATTCTTTATTTACATGATCTGTCCCATAGTCATGCATTGTTTTGCACCCCAAATTTTTTATTGTTCATAGCAGCATGGTCAGCTTTCTTCTTGATCTATAGATGAGGCTCAGGCACTATCCCATTTATACCAATAACCAGTGTATAACTACTTAAGGAAAACATAAAAACTTCATCTTCTTTCCTTTTATTTCTTATGTGAATCTCCCGTCTTCCATTCTCTTTTATAATTGAGAATGTCTCAATCATATGAAATTAGTTACCAGAATTAACACAATTTAGACTATCTTCCTGATTCCTTAAACCCCTTTACTGAAGTATACTCATGAATAATACTTTAAAATATGGGGGAATAGAAACCATGAACTTTTTACCTTTTTAAACTATTTATCCATATCTCCAAAGTAGAACATTAAACCATTTTAAGATATGTCTCATTCCCAAGTAGTCAGAGCTCACTCTCCAACTTTATTAAATACTATTTGAGCACAGGACACATTCTTAAACATTTTGAAAAACATTAACCCAAGATGTAGAGGCTACTGCTAGTCGTCATTCTAGAATCTGATATTTTACTCTGTATTTGAAATGAATGATTAATGTCCTAGGAAATTAGCTTTAGCAGATGTCCAGGTGCCACATCAAAAAAGTGCAATAATTATTGACAGTTTTTTAGATTAGGCATATTATTGGAAAACAACTTTATAAAGAGTGAACATTGTATACTCTAGTAAAACAGCATCACTTTAAAAATATTCATTTATGAAATCTGTTACCTATAGTTGAAGTCTTGAGTAGTGAACAAGGGACTCTAATACCAATACTCTTAATATCTGGCTATTTTAGATCCCTTAAAGGGCATAATTATTGGAAATTTAGGTATTTCACTAAAGCATGTATATAATATTGCCAACAAGAAAAGTAAATTTGAAGATTAAGGGAACTTACTTCTGCAAACTGTCTTGCGATAGTTAAGCAGAATTTAAACTCTGTTTTAAGCAGGAAACCAGAAAGATTATTTTGCAGTTGTAGAAGATTTCATAACTTATTAAAACTTATTAACATTTTGTGTTGTTTAGATATAGGCAGTTGATACATACTAACATCCCAGCCTTTTCAATATCAGGGTTAAATTATAGGAAAACTCAGTAAAATGGTACAAATCTGAAAGTTTGATGGTAGAAACTGAAGATTTAACAGAGAACTGTGTTTTACCCGAGTGCCAAAAATGCTGTGAGCCTCCTTGCACAAAATTTATACCACTTTTGCATTTTTATCTATCAGTCCAGATAGTTGTCTCCCCTCCTTCTCCCAGGACCTCTCCACCATTAAAATGCACAAACCACATGGCCGATTTCACCATTTACATTTATTTTCAAAAGTTACTACAACCAAATTAATTCTATTAGAAGAAATGTAGACAAATTCTATAAAGACTATAGATTGTGACCTAAGAAAGAAATGAGGCAAAGAACCAAACATTGAATTAAATGCTACATGGGTGACTAAGATCTGTTTCAAGTCAGTGATAATATAGCCACTTCTGGGTACTTCAGTATCAGAGATCAGTTCTCGTGGTTTAGACAGTTCCTATCTATAGCTGACTATCCTTGTCCTTGAATATGGTGTAACTGACTATTGGCTCTACAGTTTTATTGGGCCACTTAAGAAATATTTCCTTGAATAATTATTTTGAGAAAAAGTCTAAAAGTAATAAAAATAATTTTAAACACACTGTAGTAAGAAATGACTGTTGGAAAATTATGCTTTCACTTTCTACCATATTCTCAGCTATACAAAACCATTTATTTTGAAGATTTTTAGACTACTGTTAATTTGAAATCTGTTACTCTTATTGTGGAATTTGTTTTTTTAAAAAAGATGTTTCTAATTGGATTTTTAAAAGAAGAATGGAATTTGGTTGCTATTTTACAATAGAACCTAAGCTTTTTGTGGTTCTTAGTGTCCTATGTAAAACTTAGTGTCAAAGTAATCAACTTTGAGATTTTCCCTTCTATTCTGCTTTATATTAAAAGCCCATTAGAAAATGGGAACCTGGTGAATATATAATGAATTGTAAAATATTTTAATGTGTAACTTTTTCAACTGTGAAACTGACTTGATTTTTTGATGAAAACAGCTGCTGATAAAGTATTTTGTGTAAAGTGTAGTTCTTATTAATCAGGAAAATGATGACTTGATTAGACTGTATATGCCCTCTTGGATTTTATTTTAAATGGATTGGTGACTTTCACATAGGTAAAACACAGTCCATCTGTATTCTTTTTTCCATCAAAAATCGAGTGATTTGGAATTATAAAAAAATTGTGAGCAGCCTATTTGAAAGGCATCATGGAAATTTCACAGCACAATAACACGGATTTGTTTTTTCTTAATGATGTAAATCCGTTTAATTCATACTTTGATCAATAGCCCATGCTTGCCAACTCTGAAGAAATTTAATTTCCAGCAGTATTTTAAAGCTAGCCTGTTAACTTTTTCTGAATATTTAAAGTTCCTCTTTTTTCTATGTCTGCACAAACTGCAGACCTGGGCTGGACCCACATACTCAAGAGTCCACCTTAAGAAATTATTTTGATGTCCAAGACATCACTAAAATATTTAAGTTTAAAGATAATATGTGGTGTTAATAGATTGTGGTGCTTTTACTATTTAAAGACAACTTTCATACTTCAGATGTTTTTGAGAAGAGGGGAATGTGAGGGGAGGGGGCAGAACAGGGAGGAGTTGTTTGAATGAATTACATTCTTTATATCCATCCTGCTCATTTGGGGCATGTCTTTAAGAGAAGGCTGAAAGTTGTGAGAGTATATTGTATACCGTAAGAGAATCAACTCTTCATCATGGATGGGATTGTGAAGGCTGAACTATAAAATTCAGCATTGACAGCATCCTCAATTAATAATTCTTGGTGACAGAATAATACAGCTGGGCTGTTTTTTAAAATATAAACAATACCATTTTTAATTATTACATTAAAAATTGTAAATATATCTATGTGCCATGGCCTGGGAAGCCTGCTTTCTTTTTTCATAAAAATTATTTTTACTGTATGAAAAGATCATGGGGTTTAGCTCAAAATATCTGTGGTCCTGATAAAATTGGATTGGTAACTCTACCTCAGAAGGAAAATGGGAAAAAAAAATAGATGAGTCACAATTCAATACTTCAAGCTCAGAAACTGTGCAGATCACTGAATTTTAGATTTATAAAGTCAGAGTTGGCATGCCTTGTTTTTAATGATATGGAAGACCTTAAGAAAAAAACTTGGCTGAAGTTTAATCGTTGGTCCAGCCATTTGAAAAAGGCAATAGTTTGAGGAGGTTCCCGAATTCGGCATTTGAAATTCATTTTGTTCTCTCTTCTTCATTATTAGTGCATTTGGTGTGTGTATACTTGCACACAATTCTGTTTGTGTACACACTGCTTGCTTAGCCCTAGTCAAGAGGCATCTTTTATAAAAGGTGTAAAGAAATATCAAGGTTCTAAAATTCGGAAGAGTTTAGAATTTATTAGGAGTTTCCCAAGTTGGGATGTTAGTCTTTAAATAAACTTCATGCACCTATTCCACTTAAGGTTTTGCACCTCCTTTTTATTAGTGCAGTGCCATTTCTTCTGCTTGATTTTAGGTATGTTAATATTCCAGCCTTGCTAGTTAGCATAAAGTGACAGGTGTGAGCCATGAGGAAATTTTCTGACTTAATTTGTACACAACTACATATAAGAGTTTTAGTGGAGGAAAAAAATTAGTCCCTTGTGCGTATACAGTAGTTAGGTAAATGATTTTTCTACCAACAGTATACTCCATTCCTCATGTAGGTAAGTACAGAAAAGGTTTTTAAATGTATTTTTTTAGCCAGTTAAAGTCTATGAATCTATCTGCAACCTTATTTAATCTGTCACTATAATAATTTTGTGGTTATGCTAAGAACCATGTATACTTTTAGGTATTCTTATTTTTGTCAATTTTTCTAGGTTGGCAAGGAGGCAGAAAACCTTCATTGTTTCATATTAAAATATAATTAGACTAAACTTAATTCTAGTATGAATTTCCAAAATCATTATCTATTTATTTCATTTTTATTTAATTTTGTTTTTATTTCATTTTTAAAAGTCCCTTGTTCAATTTAACTTATGTTCCTAAGAGAGGTTGGAGAACTTGGCCTTCATCTGATTTCAAAAATGTTTTGAGTTTCAAATGAAGTTAATGGTTTCAGTGTGATTCAGTCCTCAGACCTAATTGGGTTGAATAAAATCTAAAAGAATATACCCTTTTGGAGCATAACATTTTAATACCTTGGGGAATGTGGCACTACCAAAAGAAGACTACTAACACGTCAGATGTTCACCTGGAAGCTTTATCAAGAAATTCGAACCACCCTTTTGGCCCCATTAATTGTAGCAAGTTTATTTCTCTATATTTTGTCATTCAGTGAATTGAAGTCCTGTGGTATACTGCATTCATTAGAAGAAAAACGTTTTTAATGTCCTTTTAATGATGGCCCAGAAAGCATTTGACACAGCAAGATGCATGTGTTACTATATTGAGAATATAGAATAATAACAGTATCACTAAATTTAAGACCTCTTCCCAGTCTTGCTGTTCCTAGCAAGAAGTTTGGCCTGTGACTGCACTTACTGTTTATGCTCATCAGAAACTGTCAATGTCTGCTTTTCTTTAACTCTGCAGTCTGTAACATCACGCTGTTTATTAAAAAAAAAAAGAAAAATTA\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Get sequences\n",
    "\"\"\"\n",
    "chrm_path = str(si.chromosome_path / si.get_chrm_file_name(gene_obj.chrm))\n",
    "\n",
    "transcript_seq = gene_obj.seq(transcript_id, chrm_path).upper()\n",
    "print(f\"{transcript_id=} contains {len(transcript_seq):,} bps.\")\n",
    "\n",
    "rna_seq = gene_obj.rna(transcript_id, chrm_path).upper()\n",
    "print(f\"\\nrna=\\n{rna_seq}\")\n",
    "\n",
    "orf_seq = gene_obj.ORF(transcript_id, chrm_path).upper()\n",
    "print(f\"\\norf=\\n{orf_seq}\")\n",
    "\n",
    "aa_seq = gene_obj.AA(transcript_id, chrm_path)\n",
    "print(f\"\\nAA=\\n{aa_seq}\")\n",
    "\n",
    "utr5_seq = gene_obj.UTR5(transcript_id, chrm_path).upper()\n",
    "print(f\"\\nutr5=\\n{utr5_seq}\")\n",
    "\n",
    "utr3_seq = gene_obj.UTR3(transcript_id, chrm_path).upper()\n",
    "print(f\"\\nutr3=\\n{utr3_seq}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exon_5 sequence (Exon_5:chr3:179203544:179203789):\n",
      "TATATAAGAAGCTGTATAATGCTTGGGAGGATGCCCAATTTGATGTTGATGGCTAAAGAAAGCCTTTATTCTCAACTGCCAATGGACTGTTTTACAATGCCATCTTATTCCAGACGCATTTCCACAGCTACACCATATATGAATGGAGAAACATCTACAAAATCCCTTTGGGTTATAAATAGTGCACTCAGAATAAAAATTCTTTGTGCAACCTACGTGAATGTAAATATTCGAGACATTGATAAG\n",
      "246 bps.\n",
      "Intron_5 sequence (Intron_5:chr3:179203790:179204502):\n",
      "GTAAAGTCAAATGCTGATGCTTATTATTTTATAGAAATTATTTTAGATAACCTTTTTCTTGCACTATACAGTAATCTGTTGACCTGTAGTATGTTTTCAGATGGTTAGGAGAACATCCAAATCTCCGAATGTAAAAATATATCAAGAATTTTACTTGAGCTTCCATCTACCTTAGCTATTATACAGCTcacagtcctttgttaataattctaatattcacaattctagctcttaaaatcaaaagttttacagaattcgtttggcagaaagacctgggccaaccttaagtgagggtttttataatctttattaaccccacttagtataaaattccggtatcttattaaagaaatattaatgtctttatgaggtactgcttcaccagctaaggaagtagtatttagtaagtacgtgtaccaatttagctttctaaaatatggaaaaactctgaattacatacctcccttaaggggattgtgggcctATATTTATGTTTTAGTAGTCTGATGTCTCCATTGTTATTAGTGGATGAAGGCAGCAACTAATTTTGGTGAAGACTCTACATCAGTATTAACGTGTTACATATGTGAAAAAAAGGAGAACCAAGCTATATCTGAACAAAAATTCCGTGGTTTTATATTTGAGTCTATCGAGTGTGTGCATATGTGTATGTTGAGTGTATACATTAGTATATACCTACTTTTTTCTTTTAG\n",
      "713 bps.\n"
     ]
    }
   ],
   "source": [
    "\"\"\"A sequence of a specific Exon or Intron\"\"\"\n",
    "name, number  = 'Exon', 5  # name can be either 'Intron' or 'Exon'. number is the 1-based Intron/Exon number.\n",
    "# ------------------------------------------------------------------------------\n",
    "n_seq, n_label = gene_obj.exon_intron_seq(name, number, transcript_id, chrm_path)\n",
    "print(f\"{name}_{number} sequence ({n_label}):\\n{n_seq}\\n{len(n_seq)} bps.\")\n",
    "\n",
    "name, number  = 'Intron', 5  # name can be either 'Intron' or 'Exon'. number is the 1-based Intron/Exon number.\n",
    "# ------------------------------------------------------------------------------\n",
    "n_seq, n_label = gene_obj.exon_intron_seq(name, number, transcript_id, chrm_path)\n",
    "print(f\"{name}_{number} sequence ({n_label}):\\n{n_seq}\\n{len(n_seq)} bps.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AA agrees with REST AA\n"
     ]
    }
   ],
   "source": [
    "# Optional: compare AA sequence with Ensembl REST API\n",
    "# Note: if you reeceive a \"HTTPError: 503 Server Error: Service Unavailable for url\", try to run again.\n",
    "import ensembl_rest_utils as erst\n",
    "import translation as tran\n",
    "\n",
    "p_id = gene_obj.transcripts_info[transcript_id]['protein_id']\n",
    "ensb_trans = erst.lookup_id(transcript_id)\n",
    "ensb_pid = ensb_trans['Translation']['id']\n",
    "assert ensb_pid == p_id, f\"{p_id=} and {ensb_pid=} mismatch !!\"\n",
    "ensb_aa = erst.sequence_endpoint_base(ensb_pid)\n",
    "\n",
    "if aa_seq is not None:\n",
    "    if ensb_aa == aa_seq:\n",
    "        print(f\"AA agrees with REST AA\")\n",
    "    else:\n",
    "        print(f\"AA and REST mismatch !!!!!!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'region': 'Intron_1', 'region_pos': 554, 'bp': 'C'}\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Perform some basic queries.\n",
    "\n",
    "Query a chromosome position.\n",
    "\"\"\"\n",
    "#chrm_pos = 55_157_663  # query a chromosome position\n",
    "#chrm_pos = 208_243_604-10 # query a chromosome position\n",
    "chrm_pos = 179_148_357+800 # query a chromosome position\n",
    "# -----------------------------------------------------\n",
    "chrm_info = gene_obj.chrm_pos_info(transcript_id, chrm_pos, chrm_path)\n",
    "print(chrm_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chrm_pos=179,199,690 --> rna_pos=676\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Chromosome position to RNA position\n",
    "\"\"\"\n",
    "#chrm_pos = 55_211_628  # for EGFR\n",
    "#chrm_pos = 208_236_229  # for IDH1\n",
    "chrm_pos = 179_199_690  # for PIK3CA\n",
    "# -----------------------------------------\n",
    "if (rna_pos := gene_obj.chrm_pos2rna_pos(transcript_id, chrm_pos)) is None:\n",
    "    print(f\"{chrm_pos=:,} is outside of the RNA (outside of {transcript_id} exons)\")\n",
    "else:\n",
    "    print(f\"{chrm_pos=:,} --> {rna_pos=:,}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'segment': 'ORF', 'pos_in_segment': 1192, 'NT': 'C', 'codon_number': 398, 'nt_in_codon': 1, 'codon': 'CGT', 'aa': 'R'}\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Query a position within an exon.\n",
    "'''\n",
    "exon_number = 7  # 1-based exon number\n",
    "nt_number = 47   # nucleotide 1-based number in the exon \n",
    "# -----------------------------------------------------------------------\n",
    "info = gene_obj.exon_nt_info(transcript_id, exon_number, nt_number, chrm_path)\n",
    "print(info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'codon': 'GCA', 'AA': 'A', 'codon_exon_pos': ['Exon_3:135', 'Exon_3:136', 'Exon_3:137'], 'codon_chromosome_pos': [179199824, 179199825, 179199826]}\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Query an amino-acid position.\n",
    "\n",
    "The 'codon_exon_pos' is a list of 3 positions, each in the form <exon number>:<pos>.\n",
    "pos in the 1-based position relative to the start of exon <exon number>. The corresponding\n",
    "chromosome position is then exon_start_coordinate+m*(pos-1), where m=1 [m=-1] for genes encoded on\n",
    "the positive [negative] DNA strand (in genes encoded on the negative strand, \n",
    "exon_start_coordinate>exon_end_coordinate).\n",
    "'''\n",
    "aa_number = 163 #1028 # 78 # 397  # 1-based AA position in the ORF\n",
    "# --------------------------------------------------\n",
    "aa_info = gene_obj.aa_exon_info(transcript_id, aa_number, chrm_path)\n",
    "print(aa_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mut_id='PIK3CA:chr3:Missense_Mutation:179218306:C:A' --> aa_mut='Q546K'\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Converts a DNA mutation to the corresponding AA mutation.\n",
    "\"\"\"\n",
    "#mut_id = 'EGFR:chr7:Missense_Mutation:55152609:G:C'\n",
    "#mut_id = 'IDH1:chr2:Missense_Mutation:208248455:C:T'\n",
    "mut_id = 'PIK3CA:chr3:Missense_Mutation:179218306:C:A' #'PIK3CA:chr3:Missense_Mutation:179234297:A:G'\n",
    "\n",
    "mut_id_cols = ['Gene_name', 'Chromosome', 'Variant_Classification', \n",
    "               'Start_Position', 'Reference_Allele', 'Tumor_Seq_Allele2']\n",
    "mut_id_sep = ':'\n",
    "# ==================================================================================\n",
    "\n",
    "mut_info = dict(zip(mut_id_cols, mut_id.split(mut_id_sep)))\n",
    "ref_a, mut_a, pos = mut_info['Reference_Allele'], mut_info['Tumor_Seq_Allele2'], int(mut_info['Start_Position'])\n",
    "\n",
    "aa_mut = gene_obj.DNA_SNP_mut_to_AA_mut(ref_a, mut_a, pos, transcript_id, chrm_path)\n",
    "print(f\"{mut_id=} --> {aa_mut=}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aa_mut='Q546K' corresponds to:\n",
      "{'AAA': {'start_pos': 179218306, 'reference_allele': 'CAG', 'alternative_allele': 'AAA'}, 'AAG': {'start_pos': 179218306, 'reference_allele': 'C', 'alternative_allele': 'A'}}.\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Converts a AA mutation to (all possible) DNA mutation.\n",
    "\"\"\"\n",
    "dna_all_muts = gene_obj.AA_mut_to_DNA_SNP_mut(aa_mut, transcript_id, chrm_path) # type: ignore\n",
    "print(f\"{aa_mut=} corresponds to:\\n{dna_all_muts}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example - Specifying GFF3 file or a Corresponding DataFrame\n",
    "The `Gene_cls` loads internally the latest GFF3 file into a dataframe. This load takes few seconds to complete. \n",
    "It is possible to specify a GFF3 file, or instead to provide a dataframe of the GFF3 file, when instatntiating the annotation class.\n",
    "\n",
    "For example, in cases where annotation is needed for multiple genes, it is more efficient to load the GFF3 file into a dataframe once, and then instantiate multiple annotation classes with the dataframe. Below are some examples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading GFF3 file into a dataframe, and returning a tuple containing it, and a subset of it, which contains only rows with Type values that are genes\n",
    "gff3_dfs = gaut.ensembl_gff3_df()  # analyzing genes, ncRNA_gene and pseudogene\n",
    "# gff3_dfs = gaut.ensembl_gff3_df(gene_type_values=['gene'])  # analyzing only genes (excluding ncRNA_gene and pseudogene). This slightly reduces processing times."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene name=EGFR, Gene ID=ENSG00000146648\n",
      "Source=Input DataFrame (ID=6053011280, size=(3415686, 9)). (Gene type DataFrame (ID=4546887504, size=(62754, 11)))\n",
      "Positive strand=True.\n",
      "Type=protein_coding, version=21.\n",
      "Description=epidermal growth factor receptor [Source:HGNC Symbol%3BAcc:HGNC:3236]\n",
      "Gene region=chr7:55,019,017-55,211,628.\n",
      "11 transcripts.\n",
      "Gene name=BRCA1, Gene ID=ENSG00000012048\n",
      "Source=Input DataFrame (ID=6053011280, size=(3415686, 9)). (Gene type DataFrame (ID=4546887504, size=(62754, 11)))\n",
      "Positive strand=False.\n",
      "Type=protein_coding, version=25.\n",
      "Description=BRCA1 DNA repair associated [Source:HGNC Symbol%3BAcc:HGNC:1100]\n",
      "Gene region=chr17:43,044,295-43,170,245.\n",
      "37 transcripts.\n"
     ]
    }
   ],
   "source": [
    "# now instantiating using the dataframe (much faster)\n",
    "g_a1 = gaut.Gene_cls('EGFR', gff3_dfs)\n",
    "g_a1.info()\n",
    "g_a2 = gaut.Gene_cls('BRCA1', gff3_dfs)\n",
    "g_a2.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene name=EGFR, Gene ID=ENSG00000146648\n",
      "Source=Input DataFrame (ID=6053011280, size=(3415686, 9)). (Gene type DataFrame (ID=4546887504, size=(62754, 11)))\n",
      "Positive strand=True.\n",
      "Type=protein_coding, version=21.\n",
      "Description=epidermal growth factor receptor [Source:HGNC Symbol%3BAcc:HGNC:3236]\n",
      "Gene region=chr7:55,019,017-55,211,628.\n",
      "11 transcripts.\n",
      "Gene name=BRCA1, Gene ID=ENSG00000012048\n",
      "Source=Input DataFrame (ID=6053011280, size=(3415686, 9)). (Gene type DataFrame (ID=4546887504, size=(62754, 11)))\n",
      "Positive strand=False.\n",
      "Type=protein_coding, version=25.\n",
      "Description=BRCA1 DNA repair associated [Source:HGNC Symbol%3BAcc:HGNC:1100]\n",
      "Gene region=chr17:43,044,295-43,170,245.\n",
      "37 transcripts.\n"
     ]
    }
   ],
   "source": [
    "# Another option\n",
    "g_a1 = gaut.Gene_gff3_cls('EGFR', gff3_dfs)\n",
    "g_a1.info()\n",
    "g_a2 = gaut.Gene_gff3_cls('BRCA1', gff3_dfs)\n",
    "g_a2.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /Users/yoramzarai/work/mystuff/Ramot/Projects/Cancer_mut/Data/Ensembl_gene_annotation/Homo_sapiens.GRCh38.109.gff3.gz to a dataframe...Done.\n",
      "Gene name=EGFR, Gene ID=ENSG00000146648\n",
      "Source=/Users/yoramzarai/work/mystuff/Ramot/Projects/Cancer_mut/Data/Ensembl_gene_annotation/Homo_sapiens.GRCh38.109.gff3.gz\n",
      "Positive strand=True.\n",
      "Type=protein_coding, version=21.\n",
      "Description=epidermal growth factor receptor [Source:HGNC Symbol%3BAcc:HGNC:3236]\n",
      "Gene region=chr7:55,019,017-55,211,628.\n",
      "11 transcripts.\n"
     ]
    }
   ],
   "source": [
    "# another example is to instantiate by specifying your GFF3 file\n",
    "mygff3_file: pathlib.Path = pathlib.Path(si.ROOT_path) / 'Data' / 'Ensembl_gene_annotation' / 'Homo_sapiens.GRCh38.109.gff3.gz'\n",
    "# -----------------------------------------------------------------------------------------------------------------------------\n",
    "g_af = gaut.Gene_cls('EGFR', mygff3_file)\n",
    "g_af.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /Users/yoramzarai/work/mystuff/Ramot/Projects/Cancer_mut/Data/Ensembl_gene_annotation/Homo_sapiens.GRCh38.109.gff3.gz to a dataframe...Done.\n",
      "Gene name=EGFR, Gene ID=ENSG00000146648\n",
      "Source=/Users/yoramzarai/work/mystuff/Ramot/Projects/Cancer_mut/Data/Ensembl_gene_annotation/Homo_sapiens.GRCh38.109.gff3.gz\n",
      "Positive strand=True.\n",
      "Type=protein_coding, version=21.\n",
      "Description=epidermal growth factor receptor [Source:HGNC Symbol%3BAcc:HGNC:3236]\n",
      "Gene region=chr7:55,019,017-55,211,628.\n",
      "11 transcripts.\n"
     ]
    }
   ],
   "source": [
    "# Another option\n",
    "g_af = gaut.Gene_gff3_cls('EGFR', mygff3_file)\n",
    "g_af.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading GFF3 file to a dataframe...\n",
      "Instantiating...\n",
      "Gene name=EGFR, Gene ID=ENSG00000146648\n",
      "Source=Input DataFrame (ID=6018778896, size=(3414011, 9)). (Gene type DataFrame (ID=12947940048, size=(62710, 11)))\n",
      "Positive strand=True.\n",
      "Type=protein_coding, version=21.\n",
      "Description=epidermal growth factor receptor [Source:HGNC Symbol%3BAcc:HGNC:3236]\n",
      "Gene region=chr7:55,019,017-55,211,628.\n",
      "11 transcripts.\n"
     ]
    }
   ],
   "source": [
    "# finally, you can load your GFF3 file into a dataframe, and instantiate with that dataframe (and its gene subset dataframe)\n",
    "print(\"Loading GFF3 file to a dataframe...\")\n",
    "mygff3_dfs = gaut.ensembl_gff3_df(mygff3_file)\n",
    "print(f\"Instantiating...\")\n",
    "g_ad = gaut.Gene_cls('EGFR', mygff3_dfs)\n",
    "g_ad.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene name=EGFR, Gene ID=ENSG00000146648\n",
      "Source=Input DataFrame (ID=6016522640, size=(3414011, 9)). (Gene type DataFrame (ID=6101748304, size=(62710, 11)))\n",
      "Positive strand=True.\n",
      "Type=protein_coding, version=21.\n",
      "Description=epidermal growth factor receptor [Source:HGNC Symbol%3BAcc:HGNC:3236]\n",
      "Gene region=chr7:55,019,017-55,211,628.\n",
      "11 transcripts.\n"
     ]
    }
   ],
   "source": [
    "# Another option\n",
    "mygff3_dfs = gaut.ensembl_gff3_df(mygff3_file)\n",
    "g_ad = gaut.Gene_gff3_cls('EGFR', mygff3_dfs)\n",
    "g_ad.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main Interfaces"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_list, intron_list = [1, 2, 3 ,4], [5, 6, 7, 8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import API\n",
    "import Gene_annotation as gaut\n",
    "\n",
    "# gene annotation object\n",
    "gene_obj = gaut.Gene_cls(gene)\n",
    "\n",
    "# info\n",
    "gene_obj.info()\n",
    "gene_obj.transcript_info(transcript_id, verbose=True)\n",
    "\n",
    "# maps\n",
    "gene_obj.exon_intron_map(transcript_id)\n",
    "gene_obj.exon_map(transcript_id)\n",
    "\n",
    "# sequences\n",
    "gene_obj.seq(transcript_id, chrm_path)\n",
    "gene_obj.rna(transcript_id, chrm_path)\n",
    "gene_obj.ORF(transcript_id, chrm_path)\n",
    "gene_obj.AA(transcript_id, chrm_path)\n",
    "gene_obj.UTR5(transcript_id, chrm_path)\n",
    "gene_obj.UTR3(transcript_id, chrm_path)\n",
    "gene_obj.exon_intron_seq(name, number, transcript_id, chrm_path)\n",
    "gene_obj.modified_transcript(exon_list, intron_list, transcript_id, chrm_path)\n",
    "\n",
    "# query\n",
    "gene_obj.chrm_pos_info(transcript_id, chrm_pos, chrm_path)\n",
    "gene_obj.exon_nt_info(transcript_id, exon_number, nt_number, chrm_path)\n",
    "gene_obj.aa_exon_info(transcript_id, aa_number, chrm_path)\n",
    "gene_obj.DNA_SNP_mut_to_AA_mut(ref_a, mut_a, pos, transcript_id, chrm_path)\n",
    "gene_obj.AA_mut_to_DNA_SNP_mut(aa_mut, transcript_id, chrm_path)                          # type: ignore"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv311",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
