{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene Annotation Usage Example\n",
    "This notebook provides gene annotation examples, using the `Gene_cls` or `Gene_gff3_cls` classes from `Gene_annotation.py`.\n",
    "\n",
    "The underlying data is the Ensembl GFF3 annotation file from [Ensembl gene annotation GFF3 file](https://useast.ensembl.org/Homo_sapiens/Info/Index).\n",
    "\n",
    "Requirements: Python >= 3.10, Pandas, and Numpy. In addition, add \n",
    "\n",
    "`export PYTHONPATH=\"$PYTHONPATH:/tamir2/yoramzar/Projects/Cancer_mut/Utils\"` \n",
    "\n",
    "to your `.bashrc`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Global configuration.\n",
    "'''\n",
    "import pathlib\n",
    "import pandas as pd\n",
    "\n",
    "# make sure PYTHONPATH is set, containing the Utils folder.\n",
    "import mysys_init as si\n",
    "\n",
    "# import API\n",
    "import Gene_annotation as gaut\n",
    "\n",
    "# set folder to create reports\n",
    "report_path = pathlib.Path(si.ROOT_path) / 'Ensembl_gene_anot' / 'Reports'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example - by Gene Name\n",
    "\n",
    "We use the gene name to instantiate a `Gene_cls` object, which provides the gene's annotation.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading /Users/yoramzarai/work/mystuff/Ramot/Projects/Cancer_mut/Data/Ensembl_gene_annotation/Homo_sapiens.GRCh38.111.gff3.gz to a dataframe...Done.\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Select a gene and a corresponding transcript.\n",
    "'''\n",
    "#gene, transcript_id = 'FLT3', 'ENST00000241453'\n",
    "\n",
    "gene, transcript_id = 'EGFR', 'ENST00000275493'\n",
    "#gene, transcript_id = 'IDH1', 'ENST00000345146'\n",
    "#gene, transcript_id = 'PIK3CA', 'ENST00000263967'\n",
    "\n",
    "#gene, transcript_id = 'ENSG00000290826', 'ENST00000642116'  # example of a ncRNA gene ()\n",
    "\n",
    "#gene, transcript_id = 'PIK3C2A', 'ENST00000533645'\n",
    "#gene, transcript_id = 'KLF16', 'ENST00000250916'\n",
    "\n",
    "#gene, transcript_id = 'ENSG00000290149', 'ENST00000476620'\n",
    "# ---------------------------------------------------------\n",
    "\n",
    "gene_obj = gaut.Gene_cls(gene)\n",
    "#gene_obj = gaut.Gene_cls(gene, verbose=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "gene='FLT3' information:\n",
      "Gene name=FLT3, Gene ID=ENSG00000122025\n",
      "Source=/Users/yoramzarai/work/mystuff/Ramot/Projects/Cancer_mut/Data/Ensembl_gene_annotation/Homo_sapiens.GRCh38.111.gff3.gz\n",
      "Positive strand=False.\n",
      "Type=protein_coding, version=15.\n",
      "Description=fms related receptor tyrosine kinase 3 [Source:HGNC Symbol%3BAcc:HGNC:3765]\n",
      "Gene region=chr13:28,003,274-28,100,592.\n",
      "2 transcripts.\n",
      "\n",
      "FLT3 contains 2 transcripts.\n",
      "FLT3 transcript IDs:\n",
      "1. ENST00000241453\n",
      "2. ENST00000380987\n",
      "\n",
      "Gene start = 28,003,274, gene end = 28,100,592\n",
      "\n",
      "transcript_id='ENST00000241453': start site (TSS) = 28,100,576, end site (TES) = 28,003,274.\n",
      "\n",
      "First bp of the Start codon located in chr13:28,100,510, and is bp #67 in the RNA.\n",
      "First bp of the Stop codon located in chr13:28,004,054, and is bp #3046 in the RNA.\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "General information\n",
    "\"\"\"\n",
    "print(f\"\\n{gene=} information:\")\n",
    "gene_obj.info()\n",
    "\n",
    "print(f\"\\n{gene} contains {len(gene_obj)} transcripts.\")\n",
    "gene_obj.show_all_transcript_IDs()\n",
    "\n",
    "print(f\"\\nGene start = {gene_obj.gene_start:,}, gene end = {gene_obj.gene_end:,}\")\n",
    "\n",
    "(ts, te) = gene_obj.transcript_boundaries(transcript_id)\n",
    "print(f\"\\n{transcript_id=}: start site (TSS) = {ts:,}, end site (TES) = {te:,}.\")\n",
    "\n",
    "# start and stop codons positions\n",
    "if (codons_info := gene_obj.start_and_stop_codons_pos(transcript_id)) is not None:\n",
    "    print(f\"\\nFirst bp of the Start codon located in chr{gene_obj.chrm}:{codons_info[0][0]:,}, and is bp #{codons_info[0][1]} in the RNA.\")\n",
    "    print(f\"First bp of the Stop codon located in chr{gene_obj.chrm}:{codons_info[1][0]:,}, and is bp #{codons_info[1][1]} in the RNA.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detailed transcript_id='ENST00000241453' information:\n",
      "ENST00000241453 information:\n",
      "Name=FLT3-201, ID=ENST00000241453.12, version=12\n",
      "Type=mRNA, biotype=protein_coding\n",
      "Start=28,100,576, end=28,003,274\n",
      "protein ID = ENSP00000241453\n",
      "ccdsid=CCDS31953.1\n",
      "ORF start offset = 66, ORF end offset = 3,047 (0-based offsets from the beginning of the RNA)\n",
      "ORF start chromosome position = 28,100,510, ORF end chromosome position = 28,004,052\n",
      "Exon regions:\n",
      "\t1. 28,100,576 - 28,100,468, ID=ENSE00003896396, phase=-1, end phase=1\n",
      "\t2. 28,070,612 - 28,070,491, ID=ENSE00000906615, phase=1, end phase=0\n",
      "\t3. 28,062,069 - 28,061,867, ID=ENSE00000906614, phase=0, end phase=2\n",
      "\t4. 28,057,462 - 28,057,347, ID=ENSE00000906613, phase=2, end phase=1\n",
      "\t5. 28,052,674 - 28,052,545, ID=ENSE00000906612, phase=1, end phase=2\n",
      "\t6. 28,050,222 - 28,050,095, ID=ENSE00000906611, phase=2, end phase=1\n",
      "\t7. 28,049,774 - 28,049,635, ID=ENSE00000906610, phase=1, end phase=0\n",
      "\t8. 28,049,537 - 28,049,384, ID=ENSE00000906609, phase=0, end phase=1\n",
      "\t9. 28,048,443 - 28,048,275, ID=ENSE00000906608, phase=1, end phase=2\n",
      "\t10. 28,037,288 - 28,037,185, ID=ENSE00000827401, phase=2, end phase=1\n",
      "\t11. 28,036,043 - 28,035,935, ID=ENSE00000827402, phase=1, end phase=2\n",
      "\t12. 28,035,673 - 28,035,495, ID=ENSE00000827403, phase=2, end phase=1\n",
      "\t13. 28,034,407 - 28,034,301, ID=ENSE00000827404, phase=1, end phase=0\n",
      "\t14. 28,034,214 - 28,034,082, ID=ENSE00000827405, phase=0, end phase=1\n",
      "\t15. 28,033,991 - 28,033,887, ID=ENSE00000827406, phase=1, end phase=1\n",
      "\t16. 28,028,288 - 28,028,178, ID=ENSE00000827407, phase=1, end phase=1\n",
      "\t17. 28,027,241 - 28,027,088, ID=ENSE00000827408, phase=1, end phase=2\n",
      "\t18. 28,024,943 - 28,024,861, ID=ENSE00003565636, phase=2, end phase=1\n",
      "\t19. 28,023,477 - 28,023,350, ID=ENSE00003519436, phase=1, end phase=0\n",
      "\t20. 28,018,589 - 28,018,467, ID=ENSE00003598146, phase=0, end phase=0\n",
      "\t21. 28,015,701 - 28,015,590, ID=ENSE00003552851, phase=0, end phase=1\n",
      "\t22. 28,015,256 - 28,015,157, ID=ENSE00003692016, phase=1, end phase=2\n",
      "\t23. 28,014,557 - 28,014,452, ID=ENSE00003513156, phase=2, end phase=0\n",
      "\t24. 28,004,174 - 28,003,274, ID=ENSE00001487116, phase=0, end phase=-1\n",
      "CDS regions:\n",
      "\t1. 28,100,510 - 28,100,468, phase=0\n",
      "\t2. 28,070,612 - 28,070,491, phase=2\n",
      "\t3. 28,062,069 - 28,061,867, phase=0\n",
      "\t4. 28,057,462 - 28,057,347, phase=1\n",
      "\t5. 28,052,674 - 28,052,545, phase=2\n",
      "\t6. 28,050,222 - 28,050,095, phase=1\n",
      "\t7. 28,049,774 - 28,049,635, phase=2\n",
      "\t8. 28,049,537 - 28,049,384, phase=0\n",
      "\t9. 28,048,443 - 28,048,275, phase=2\n",
      "\t10. 28,037,288 - 28,037,185, phase=1\n",
      "\t11. 28,036,043 - 28,035,935, phase=2\n",
      "\t12. 28,035,673 - 28,035,495, phase=1\n",
      "\t13. 28,034,407 - 28,034,301, phase=2\n",
      "\t14. 28,034,214 - 28,034,082, phase=0\n",
      "\t15. 28,033,991 - 28,033,887, phase=2\n",
      "\t16. 28,028,288 - 28,028,178, phase=2\n",
      "\t17. 28,027,241 - 28,027,088, phase=2\n",
      "\t18. 28,024,943 - 28,024,861, phase=1\n",
      "\t19. 28,023,477 - 28,023,350, phase=2\n",
      "\t20. 28,018,589 - 28,018,467, phase=0\n",
      "\t21. 28,015,701 - 28,015,590, phase=0\n",
      "\t22. 28,015,256 - 28,015,157, phase=2\n",
      "\t23. 28,014,557 - 28,014,452, phase=1\n",
      "\t24. 28,004,174 - 28,004,052, phase=0\n",
      "5UTR regions:\n",
      "\t1. 28,100,576 - 28,100,511\n",
      "3UTR regions:\n",
      "\t1. 28,004,051 - 28,003,274\n"
     ]
    }
   ],
   "source": [
    "# get a detailed information about a transcript\n",
    "print(f\"Detailed {transcript_id=} information:\")\n",
    "gene_obj.transcript_info(transcript_id, verbose=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Regarding the exon phases see [this](https://blog.gene-test.com/coding-phases-frames/). Basically:\n",
    "1. exon end phase is the exon phase of the next exon.\n",
    "2. Conversion from exon phase to CDS phase: 0-->0, 1-->2, 2-->1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Exon-intron map for transcript_id='ENST00000241453':\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>region</th>\n",
       "      <th>region_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Exon 1</td>\n",
       "      <td>(28100576, 28100468)</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Intron 1</td>\n",
       "      <td>(28100467, 28070613)</td>\n",
       "      <td>29855</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Exon 2</td>\n",
       "      <td>(28070612, 28070491)</td>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Intron 2</td>\n",
       "      <td>(28070490, 28062070)</td>\n",
       "      <td>8421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Exon 3</td>\n",
       "      <td>(28062069, 28061867)</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Intron 3</td>\n",
       "      <td>(28061866, 28057463)</td>\n",
       "      <td>4404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Exon 4</td>\n",
       "      <td>(28057462, 28057347)</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Intron 4</td>\n",
       "      <td>(28057346, 28052675)</td>\n",
       "      <td>4672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Exon 5</td>\n",
       "      <td>(28052674, 28052545)</td>\n",
       "      <td>130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Intron 5</td>\n",
       "      <td>(28052544, 28050223)</td>\n",
       "      <td>2322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Exon 6</td>\n",
       "      <td>(28050222, 28050095)</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Intron 6</td>\n",
       "      <td>(28050094, 28049775)</td>\n",
       "      <td>320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Exon 7</td>\n",
       "      <td>(28049774, 28049635)</td>\n",
       "      <td>140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Intron 7</td>\n",
       "      <td>(28049634, 28049538)</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Exon 8</td>\n",
       "      <td>(28049537, 28049384)</td>\n",
       "      <td>154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Intron 8</td>\n",
       "      <td>(28049383, 28048444)</td>\n",
       "      <td>940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Exon 9</td>\n",
       "      <td>(28048443, 28048275)</td>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Intron 9</td>\n",
       "      <td>(28048274, 28037289)</td>\n",
       "      <td>10986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Exon 10</td>\n",
       "      <td>(28037288, 28037185)</td>\n",
       "      <td>104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Intron 10</td>\n",
       "      <td>(28037184, 28036044)</td>\n",
       "      <td>1141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Exon 11</td>\n",
       "      <td>(28036043, 28035935)</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Intron 11</td>\n",
       "      <td>(28035934, 28035674)</td>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Exon 12</td>\n",
       "      <td>(28035673, 28035495)</td>\n",
       "      <td>179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Intron 12</td>\n",
       "      <td>(28035494, 28034408)</td>\n",
       "      <td>1087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Exon 13</td>\n",
       "      <td>(28034407, 28034301)</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Intron 13</td>\n",
       "      <td>(28034300, 28034215)</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Exon 14</td>\n",
       "      <td>(28034214, 28034082)</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Intron 14</td>\n",
       "      <td>(28034081, 28033992)</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Exon 15</td>\n",
       "      <td>(28033991, 28033887)</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Intron 15</td>\n",
       "      <td>(28033886, 28028289)</td>\n",
       "      <td>5598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Exon 16</td>\n",
       "      <td>(28028288, 28028178)</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Intron 16</td>\n",
       "      <td>(28028177, 28027242)</td>\n",
       "      <td>936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Exon 17</td>\n",
       "      <td>(28027241, 28027088)</td>\n",
       "      <td>154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Intron 17</td>\n",
       "      <td>(28027087, 28024944)</td>\n",
       "      <td>2144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Exon 18</td>\n",
       "      <td>(28024943, 28024861)</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Intron 18</td>\n",
       "      <td>(28024860, 28023478)</td>\n",
       "      <td>1383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Exon 19</td>\n",
       "      <td>(28023477, 28023350)</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Intron 19</td>\n",
       "      <td>(28023349, 28018590)</td>\n",
       "      <td>4760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Exon 20</td>\n",
       "      <td>(28018589, 28018467)</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Intron 20</td>\n",
       "      <td>(28018466, 28015702)</td>\n",
       "      <td>2765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Exon 21</td>\n",
       "      <td>(28015701, 28015590)</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Intron 21</td>\n",
       "      <td>(28015589, 28015257)</td>\n",
       "      <td>333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Exon 22</td>\n",
       "      <td>(28015256, 28015157)</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Intron 22</td>\n",
       "      <td>(28015156, 28014558)</td>\n",
       "      <td>599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Exon 23</td>\n",
       "      <td>(28014557, 28014452)</td>\n",
       "      <td>106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Intron 23</td>\n",
       "      <td>(28014451, 28004175)</td>\n",
       "      <td>10277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Exon 24</td>\n",
       "      <td>(28004174, 28003274)</td>\n",
       "      <td>901</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         name                region  region_size\n",
       "0      Exon 1  (28100576, 28100468)          109\n",
       "1    Intron 1  (28100467, 28070613)        29855\n",
       "2      Exon 2  (28070612, 28070491)          122\n",
       "3    Intron 2  (28070490, 28062070)         8421\n",
       "4      Exon 3  (28062069, 28061867)          203\n",
       "5    Intron 3  (28061866, 28057463)         4404\n",
       "6      Exon 4  (28057462, 28057347)          116\n",
       "7    Intron 4  (28057346, 28052675)         4672\n",
       "8      Exon 5  (28052674, 28052545)          130\n",
       "9    Intron 5  (28052544, 28050223)         2322\n",
       "10     Exon 6  (28050222, 28050095)          128\n",
       "11   Intron 6  (28050094, 28049775)          320\n",
       "12     Exon 7  (28049774, 28049635)          140\n",
       "13   Intron 7  (28049634, 28049538)           97\n",
       "14     Exon 8  (28049537, 28049384)          154\n",
       "15   Intron 8  (28049383, 28048444)          940\n",
       "16     Exon 9  (28048443, 28048275)          169\n",
       "17   Intron 9  (28048274, 28037289)        10986\n",
       "18    Exon 10  (28037288, 28037185)          104\n",
       "19  Intron 10  (28037184, 28036044)         1141\n",
       "20    Exon 11  (28036043, 28035935)          109\n",
       "21  Intron 11  (28035934, 28035674)          261\n",
       "22    Exon 12  (28035673, 28035495)          179\n",
       "23  Intron 12  (28035494, 28034408)         1087\n",
       "24    Exon 13  (28034407, 28034301)          107\n",
       "25  Intron 13  (28034300, 28034215)           86\n",
       "26    Exon 14  (28034214, 28034082)          133\n",
       "27  Intron 14  (28034081, 28033992)           90\n",
       "28    Exon 15  (28033991, 28033887)          105\n",
       "29  Intron 15  (28033886, 28028289)         5598\n",
       "30    Exon 16  (28028288, 28028178)          111\n",
       "31  Intron 16  (28028177, 28027242)          936\n",
       "32    Exon 17  (28027241, 28027088)          154\n",
       "33  Intron 17  (28027087, 28024944)         2144\n",
       "34    Exon 18  (28024943, 28024861)           83\n",
       "35  Intron 18  (28024860, 28023478)         1383\n",
       "36    Exon 19  (28023477, 28023350)          128\n",
       "37  Intron 19  (28023349, 28018590)         4760\n",
       "38    Exon 20  (28018589, 28018467)          123\n",
       "39  Intron 20  (28018466, 28015702)         2765\n",
       "40    Exon 21  (28015701, 28015590)          112\n",
       "41  Intron 21  (28015589, 28015257)          333\n",
       "42    Exon 22  (28015256, 28015157)          100\n",
       "43  Intron 22  (28015156, 28014558)          599\n",
       "44    Exon 23  (28014557, 28014452)          106\n",
       "45  Intron 23  (28014451, 28004175)        10277\n",
       "46    Exon 24  (28004174, 28003274)          901"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# exon-intron table\n",
    "print(f'\\nExon-intron map for {transcript_id=}:')\n",
    "display(gene_obj.exon_intron_map(transcript_id))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Exon map for transcript_id='ENST00000241453':\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_region</th>\n",
       "      <th>exon_size</th>\n",
       "      <th>mRNA_NT_region</th>\n",
       "      <th>ORF_NT_region</th>\n",
       "      <th>ORF_AA_region</th>\n",
       "      <th>next_exon_frame_alignment</th>\n",
       "      <th>exon_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>(28100576, 28100468)</td>\n",
       "      <td>109</td>\n",
       "      <td>[1, 109]</td>\n",
       "      <td>[1, 43]</td>\n",
       "      <td>[1, 15]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003896396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>(28070612, 28070491)</td>\n",
       "      <td>122</td>\n",
       "      <td>[110, 231]</td>\n",
       "      <td>[44, 165]</td>\n",
       "      <td>[15, 55]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00000906615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>(28062069, 28061867)</td>\n",
       "      <td>203</td>\n",
       "      <td>[232, 434]</td>\n",
       "      <td>[166, 368]</td>\n",
       "      <td>[56, 123]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00000906614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>(28057462, 28057347)</td>\n",
       "      <td>116</td>\n",
       "      <td>[435, 550]</td>\n",
       "      <td>[369, 484]</td>\n",
       "      <td>[123, 162]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000906613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>(28052674, 28052545)</td>\n",
       "      <td>130</td>\n",
       "      <td>[551, 680]</td>\n",
       "      <td>[485, 614]</td>\n",
       "      <td>[162, 205]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00000906612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>(28050222, 28050095)</td>\n",
       "      <td>128</td>\n",
       "      <td>[681, 808]</td>\n",
       "      <td>[615, 742]</td>\n",
       "      <td>[205, 248]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000906611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>(28049774, 28049635)</td>\n",
       "      <td>140</td>\n",
       "      <td>[809, 948]</td>\n",
       "      <td>[743, 882]</td>\n",
       "      <td>[248, 294]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00000906610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>(28049537, 28049384)</td>\n",
       "      <td>154</td>\n",
       "      <td>[949, 1102]</td>\n",
       "      <td>[883, 1036]</td>\n",
       "      <td>[295, 346]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000906609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>(28048443, 28048275)</td>\n",
       "      <td>169</td>\n",
       "      <td>[1103, 1271]</td>\n",
       "      <td>[1037, 1205]</td>\n",
       "      <td>[346, 402]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00000906608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>(28037288, 28037185)</td>\n",
       "      <td>104</td>\n",
       "      <td>[1272, 1375]</td>\n",
       "      <td>[1206, 1309]</td>\n",
       "      <td>[402, 437]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000827401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>(28036043, 28035935)</td>\n",
       "      <td>109</td>\n",
       "      <td>[1376, 1484]</td>\n",
       "      <td>[1310, 1418]</td>\n",
       "      <td>[437, 473]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00000827402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>(28035673, 28035495)</td>\n",
       "      <td>179</td>\n",
       "      <td>[1485, 1663]</td>\n",
       "      <td>[1419, 1597]</td>\n",
       "      <td>[473, 533]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000827403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>(28034407, 28034301)</td>\n",
       "      <td>107</td>\n",
       "      <td>[1664, 1770]</td>\n",
       "      <td>[1598, 1704]</td>\n",
       "      <td>[533, 568]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00000827404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>(28034214, 28034082)</td>\n",
       "      <td>133</td>\n",
       "      <td>[1771, 1903]</td>\n",
       "      <td>[1705, 1837]</td>\n",
       "      <td>[569, 613]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000827405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>(28033991, 28033887)</td>\n",
       "      <td>105</td>\n",
       "      <td>[1904, 2008]</td>\n",
       "      <td>[1838, 1942]</td>\n",
       "      <td>[613, 648]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000827406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>(28028288, 28028178)</td>\n",
       "      <td>111</td>\n",
       "      <td>[2009, 2119]</td>\n",
       "      <td>[1943, 2053]</td>\n",
       "      <td>[648, 685]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00000827407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>(28027241, 28027088)</td>\n",
       "      <td>154</td>\n",
       "      <td>[2120, 2273]</td>\n",
       "      <td>[2054, 2207]</td>\n",
       "      <td>[685, 736]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00000827408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>(28024943, 28024861)</td>\n",
       "      <td>83</td>\n",
       "      <td>[2274, 2356]</td>\n",
       "      <td>[2208, 2290]</td>\n",
       "      <td>[736, 764]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003565636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>(28023477, 28023350)</td>\n",
       "      <td>128</td>\n",
       "      <td>[2357, 2484]</td>\n",
       "      <td>[2291, 2418]</td>\n",
       "      <td>[764, 806]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00003519436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>(28018589, 28018467)</td>\n",
       "      <td>123</td>\n",
       "      <td>[2485, 2607]</td>\n",
       "      <td>[2419, 2541]</td>\n",
       "      <td>[807, 847]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00003598146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>(28015701, 28015590)</td>\n",
       "      <td>112</td>\n",
       "      <td>[2608, 2719]</td>\n",
       "      <td>[2542, 2653]</td>\n",
       "      <td>[848, 885]</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003552851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>(28015256, 28015157)</td>\n",
       "      <td>100</td>\n",
       "      <td>[2720, 2819]</td>\n",
       "      <td>[2654, 2753]</td>\n",
       "      <td>[885, 918]</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00003692016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>23</td>\n",
       "      <td>(28014557, 28014452)</td>\n",
       "      <td>106</td>\n",
       "      <td>[2820, 2925]</td>\n",
       "      <td>[2754, 2859]</td>\n",
       "      <td>[918, 953]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00003513156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>24</td>\n",
       "      <td>(28004174, 28003274)</td>\n",
       "      <td>901</td>\n",
       "      <td>[2926, 3826]</td>\n",
       "      <td>[2860, 2982]</td>\n",
       "      <td>[954, 993]</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSE00001487116</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    exon_number           exon_region  exon_size mRNA_NT_region ORF_NT_region  \\\n",
       "0             1  (28100576, 28100468)        109       [1, 109]       [1, 43]   \n",
       "1             2  (28070612, 28070491)        122     [110, 231]     [44, 165]   \n",
       "2             3  (28062069, 28061867)        203     [232, 434]    [166, 368]   \n",
       "3             4  (28057462, 28057347)        116     [435, 550]    [369, 484]   \n",
       "4             5  (28052674, 28052545)        130     [551, 680]    [485, 614]   \n",
       "5             6  (28050222, 28050095)        128     [681, 808]    [615, 742]   \n",
       "6             7  (28049774, 28049635)        140     [809, 948]    [743, 882]   \n",
       "7             8  (28049537, 28049384)        154    [949, 1102]   [883, 1036]   \n",
       "8             9  (28048443, 28048275)        169   [1103, 1271]  [1037, 1205]   \n",
       "9            10  (28037288, 28037185)        104   [1272, 1375]  [1206, 1309]   \n",
       "10           11  (28036043, 28035935)        109   [1376, 1484]  [1310, 1418]   \n",
       "11           12  (28035673, 28035495)        179   [1485, 1663]  [1419, 1597]   \n",
       "12           13  (28034407, 28034301)        107   [1664, 1770]  [1598, 1704]   \n",
       "13           14  (28034214, 28034082)        133   [1771, 1903]  [1705, 1837]   \n",
       "14           15  (28033991, 28033887)        105   [1904, 2008]  [1838, 1942]   \n",
       "15           16  (28028288, 28028178)        111   [2009, 2119]  [1943, 2053]   \n",
       "16           17  (28027241, 28027088)        154   [2120, 2273]  [2054, 2207]   \n",
       "17           18  (28024943, 28024861)         83   [2274, 2356]  [2208, 2290]   \n",
       "18           19  (28023477, 28023350)        128   [2357, 2484]  [2291, 2418]   \n",
       "19           20  (28018589, 28018467)        123   [2485, 2607]  [2419, 2541]   \n",
       "20           21  (28015701, 28015590)        112   [2608, 2719]  [2542, 2653]   \n",
       "21           22  (28015256, 28015157)        100   [2720, 2819]  [2654, 2753]   \n",
       "22           23  (28014557, 28014452)        106   [2820, 2925]  [2754, 2859]   \n",
       "23           24  (28004174, 28003274)        901   [2926, 3826]  [2860, 2982]   \n",
       "\n",
       "   ORF_AA_region  next_exon_frame_alignment          exon_ID  \n",
       "0        [1, 15]                          2  ENSE00003896396  \n",
       "1       [15, 55]                          0  ENSE00000906615  \n",
       "2      [56, 123]                          1  ENSE00000906614  \n",
       "3     [123, 162]                          2  ENSE00000906613  \n",
       "4     [162, 205]                          1  ENSE00000906612  \n",
       "5     [205, 248]                          2  ENSE00000906611  \n",
       "6     [248, 294]                          0  ENSE00000906610  \n",
       "7     [295, 346]                          2  ENSE00000906609  \n",
       "8     [346, 402]                          1  ENSE00000906608  \n",
       "9     [402, 437]                          2  ENSE00000827401  \n",
       "10    [437, 473]                          1  ENSE00000827402  \n",
       "11    [473, 533]                          2  ENSE00000827403  \n",
       "12    [533, 568]                          0  ENSE00000827404  \n",
       "13    [569, 613]                          2  ENSE00000827405  \n",
       "14    [613, 648]                          2  ENSE00000827406  \n",
       "15    [648, 685]                          2  ENSE00000827407  \n",
       "16    [685, 736]                          1  ENSE00000827408  \n",
       "17    [736, 764]                          2  ENSE00003565636  \n",
       "18    [764, 806]                          0  ENSE00003519436  \n",
       "19    [807, 847]                          0  ENSE00003598146  \n",
       "20    [848, 885]                          2  ENSE00003552851  \n",
       "21    [885, 918]                          1  ENSE00003692016  \n",
       "22    [918, 953]                          0  ENSE00003513156  \n",
       "23    [954, 993]                          0  ENSE00001487116  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# exon table\n",
    "print(f'\\nExon map for {transcript_id=}:')\n",
    "display(gene_obj.exon_map(transcript_id))\n",
    "\n",
    "# if (df := gene_obj.exon_map(transcript_id)) is not None:\n",
    "#     print(f\"Exon map for {gene=}, {transcript_id=}:\")\n",
    "#     display(df.style.format(formatter={\"exon_region\": lambda x: f'({x[0]:,}, {x[1]:,})'}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "exon map can be save to excel, csv, or html files.\n",
    "'''\n",
    "gene_obj.exon_map_to_html(transcript_id, report_path / f\"{gene}_{transcript_id}_exon_map.html\")\n",
    "gene_obj.exon_map_to_excel(transcript_id, report_path / f\"{gene}_{transcript_id}_exon_map.xlsx\", usr_desc={\"General\": f\"Exon table for {gene}, {transcript_id}\", \"Source\": \"gene_annotation_example.ipynb\" })\n",
    "gene_obj.exon_map_to_csv(transcript_id, report_path / f\"{gene}_{transcript_id}_exon_map.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Get sequences\n",
    "\"\"\"\n",
    "chrm_path = str(si.chromosome_path / si.get_chrm_file_name(gene_obj.chrm))\n",
    "\n",
    "transcript_seq = gene_obj.seq(transcript_id, chrm_path).upper()\n",
    "print(f\"{transcript_id=} contains {len(transcript_seq):,} bps.\")\n",
    "\n",
    "rna_seq = gene_obj.rna(transcript_id, chrm_path).upper()\n",
    "print(f\"\\nrna=\\n{rna_seq}\")\n",
    "\n",
    "orf_seq = gene_obj.ORF(transcript_id, chrm_path).upper()\n",
    "print(f\"\\norf=\\n{orf_seq}\")\n",
    "\n",
    "aa_seq = gene_obj.AA(transcript_id, chrm_path)\n",
    "print(f\"\\nAA=\\n{aa_seq}\")\n",
    "\n",
    "utr5_seq = gene_obj.UTR5(transcript_id, chrm_path).upper()\n",
    "print(f\"\\nutr5=\\n{utr5_seq}\")\n",
    "\n",
    "utr3_seq = gene_obj.UTR3(transcript_id, chrm_path).upper()\n",
    "print(f\"\\nutr3=\\n{utr3_seq}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"A sequence of a specific Exon or Intron\"\"\"\n",
    "name, number  = 'Exon', 5  # name can be either 'Intron' or 'Exon'. number is the 1-based Intron/Exon number.\n",
    "# ------------------------------------------------------------------------------\n",
    "n_seq, n_label = gene_obj.exon_intron_seq(name, number, transcript_id, chrm_path)\n",
    "print(f\"{name}_{number} sequence ({n_label}):\\n{n_seq}\\n{len(n_seq)} bps.\")\n",
    "\n",
    "name, number  = 'Intron', 5  # name can be either 'Intron' or 'Exon'. number is the 1-based Intron/Exon number.\n",
    "# ------------------------------------------------------------------------------\n",
    "n_seq, n_label = gene_obj.exon_intron_seq(name, number, transcript_id, chrm_path)\n",
    "print(f\"{name}_{number} sequence ({n_label}):\\n{n_seq}\\n{len(n_seq)} bps.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Optional: compare AA sequence with Ensembl REST API\n",
    "# Note: if you reeceive a \"HTTPError: 503 Server Error: Service Unavailable for url\", try to run again.\n",
    "import ensembl_rest_utils as erut\n",
    "ensb_aa = erut.sequence_endpoint_base(transcript_id, seq_type='protein')\n",
    "if aa_seq is not None:\n",
    "    if ensb_aa == aa_seq:\n",
    "        print(f\"AA agrees with REST AA\")\n",
    "    else:\n",
    "        print(f\"AA and REST mismatch !!!!!!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Perform some basic queries.\n",
    "\n",
    "Query a chromosome position.\n",
    "\"\"\"\n",
    "chrm_pos = 55_157_663  # query a chromosome position\n",
    "#chrm_pos = 208_243_604-10 # query a chromosome position\n",
    "#chrm_pos = 179_148_357+800 # query a chromosome position\n",
    "# -----------------------------------------------------\n",
    "chrm_info = gene_obj.chrm_pos_info(transcript_id, chrm_pos, chrm_path)\n",
    "print(chrm_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Chromosome position to RNA position\n",
    "\"\"\"\n",
    "chrm_pos = 55_211_628  # for EGFR\n",
    "#chrm_pos = 208_236_229  # for IDH1\n",
    "#chrm_pos = 179_199_690  # for PIK3CA\n",
    "# -----------------------------------------\n",
    "if (rna_pos := gene_obj.chrm_pos2rna_pos(transcript_id, chrm_pos)) is None:\n",
    "    print(f\"{chrm_pos=:,} is outside of the RNA (outside of {transcript_id} exons)\")\n",
    "else:\n",
    "    print(f\"{chrm_pos=:,} --> {rna_pos=:,}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "RNA position to chromosome position    \n",
    "\"\"\"\n",
    "rna_pos: int = 685\n",
    "# ================\n",
    "chrm_p = gene_obj.rna_pos2chrm_pos(transcript_id, rna_pos, chrm_path)\n",
    "print(f\"{rna_pos=} --> {chrm_p=}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "RNA position to chromosome information   \n",
    "\"\"\"\n",
    "rna_pos: int = 685\n",
    "# ================\n",
    "chrm_info = gene_obj.rna_pos2chrm_info(transcript_id, rna_pos, chrm_path)\n",
    "print(chrm_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Query a position within an exon.\n",
    "'''\n",
    "exon_number = 7  # 1-based exon number\n",
    "nt_number = 47   # nucleotide 1-based number in the exon \n",
    "# -----------------------------------------------------------------------\n",
    "info = gene_obj.exon_nt_info(transcript_id, exon_number, nt_number, chrm_path)\n",
    "print(info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Query an amino-acid position.\n",
    "\n",
    "The 'codon_exon_pos' is a list of 3 positions, each in the form <exon number>:<pos>.\n",
    "pos in the 1-based position relative to the start of exon <exon number>. The corresponding\n",
    "chromosome position is then exon_start_coordinate+m*(pos-1), where m=1 [m=-1] for genes encoded on\n",
    "the positive [negative] DNA strand (in genes encoded on the negative strand, \n",
    "exon_start_coordinate>exon_end_coordinate).\n",
    "'''\n",
    "aa_number = 163 #1028 # 78 # 397  # 1-based AA position in the ORF\n",
    "# --------------------------------------------------\n",
    "aa_info = gene_obj.aa_exon_info(transcript_id, aa_number, chrm_path)\n",
    "print(aa_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Converts a DNA mutation to the corresponding AA mutation.\n",
    "\"\"\"\n",
    "mut_id = 'EGFR:chr7:Missense_Mutation:55152609:G:C'\n",
    "#mut_id = 'IDH1:chr2:Missense_Mutation:208248455:C:T'\n",
    "#mut_id = 'PIK3CA:chr3:Missense_Mutation:179218306:C:A' #'PIK3CA:chr3:Missense_Mutation:179234297:A:G'\n",
    "\n",
    "mut_id_cols = ['Gene_name', 'Chromosome', 'Variant_Classification', \n",
    "               'Start_Position', 'Reference_Allele', 'Tumor_Seq_Allele2']\n",
    "mut_id_sep = ':'\n",
    "# ==================================================================================\n",
    "\n",
    "mut_info = dict(zip(mut_id_cols, mut_id.split(mut_id_sep)))\n",
    "ref_a, mut_a, pos = mut_info['Reference_Allele'], mut_info['Tumor_Seq_Allele2'], int(mut_info['Start_Position'])\n",
    "\n",
    "aa_mut = gene_obj.DNA_SNP_mut_to_AA_mut(ref_a, mut_a, pos, transcript_id, chrm_path)\n",
    "print(f\"{mut_id=} --> {aa_mut=}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Converts a AA mutation to (all possible) DNA mutation.\n",
    "\"\"\"\n",
    "dna_all_muts = gene_obj.AA_mut_to_DNA_SNP_mut(aa_mut, transcript_id, chrm_path) # type: ignore\n",
    "print(f\"{aa_mut=} corresponds to:\\n{dna_all_muts}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example - Specifying GFF3 file or a Corresponding DataFrame\n",
    "The `Gene_cls` loads internally the latest GFF3 file into a dataframe. This load takes few seconds to complete. \n",
    "It is possible to specify a GFF3 file, or instead to provide a dataframe of the GFF3 file, when instatntiating the annotation class.\n",
    "\n",
    "For example, in cases where annotation is needed for multiple genes, it is more efficient to load the GFF3 file into a dataframe once, and then instantiate multiple annotation classes with the dataframe. Below are some examples."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading GFF3 file into a dataframe, and returning a tuple containing it, and a subset of it, which contains only rows with Type values that are genes\n",
    "gff3_dfs = gaut.ensembl_gff3_df()  # analyzing genes, ncRNA_gene and pseudogene\n",
    "# gff3_dfs = gaut.ensembl_gff3_df(gene_type_values=['gene'])  # analyzing only genes (excluding ncRNA_gene and pseudogene). This slightly reduces processing times."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now instantiating using the dataframe (much faster)\n",
    "g_a1 = gaut.Gene_cls('EGFR', gff3_dfs)\n",
    "g_a1.info()\n",
    "g_a2 = gaut.Gene_cls('BRCA1', gff3_dfs)\n",
    "g_a2.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Another option\n",
    "g_a1 = gaut.Gene_gff3_cls('EGFR', gff3_dfs)\n",
    "g_a1.info()\n",
    "g_a2 = gaut.Gene_gff3_cls('BRCA1', gff3_dfs)\n",
    "g_a2.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# another example is to instantiate by specifying your GFF3 file\n",
    "mygff3_file: pathlib.Path = pathlib.Path(si.ROOT_path) / 'Data' / 'Ensembl_gene_annotation' / 'Homo_sapiens.GRCh38.110.gff3.gz'\n",
    "# -----------------------------------------------------------------------------------------------------------------------------\n",
    "g_af = gaut.Gene_cls('EGFR', mygff3_file)\n",
    "g_af.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Another option\n",
    "g_af = gaut.Gene_gff3_cls('EGFR', mygff3_file)\n",
    "g_af.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# finally, you can load your GFF3 file into a dataframe, and instantiate with that dataframe (and its gene subset dataframe)\n",
    "print(\"Loading GFF3 file to a dataframe...\")\n",
    "mygff3_dfs = gaut.ensembl_gff3_df(mygff3_file)\n",
    "print(f\"Instantiating...\")\n",
    "g_ad = gaut.Gene_cls('EGFR', mygff3_dfs)\n",
    "g_ad.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Another option\n",
    "mygff3_dfs = gaut.ensembl_gff3_df(mygff3_file)\n",
    "g_ad = gaut.Gene_gff3_cls('EGFR', mygff3_dfs)\n",
    "g_ad.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Main Interfaces"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exon_list, intron_list = [1, 2, 3 ,4], [5, 6, 7, 8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import API\n",
    "import Gene_annotation as gaut\n",
    "\n",
    "# gene annotation object\n",
    "gene_obj = gaut.Gene_cls(gene)\n",
    "\n",
    "# info\n",
    "gene_obj.info()\n",
    "gene_obj.transcript_info(transcript_id, verbose=True)\n",
    "\n",
    "# maps\n",
    "gene_obj.exon_intron_map(transcript_id)\n",
    "gene_obj.exon_map(transcript_id)\n",
    "\n",
    "# sequences\n",
    "gene_obj.seq(transcript_id, chrm_path)\n",
    "gene_obj.rna(transcript_id, chrm_path)\n",
    "gene_obj.ORF(transcript_id, chrm_path)\n",
    "gene_obj.AA(transcript_id, chrm_path)\n",
    "gene_obj.UTR5(transcript_id, chrm_path)\n",
    "gene_obj.UTR3(transcript_id, chrm_path)\n",
    "gene_obj.exon_intron_seq(name, number, transcript_id, chrm_path)\n",
    "gene_obj.modified_transcript(exon_list, intron_list, transcript_id, chrm_path)\n",
    "\n",
    "# query\n",
    "gene_obj.chrm_pos_info(transcript_id, chrm_pos, chrm_path)\n",
    "gene_obj.rna_pos2chrm_pos(transcript_id, rna_pos, chrm_path)\n",
    "gene_obj.rna_pos2chrm_info(transcript_id, rna_pos, chrm_path)\n",
    "gene_obj.exon_nt_info(transcript_id, exon_number, nt_number, chrm_path)\n",
    "gene_obj.aa_exon_info(transcript_id, aa_number, chrm_path)\n",
    "gene_obj.DNA_SNP_mut_to_AA_mut(ref_a, mut_a, pos, transcript_id, chrm_path)\n",
    "gene_obj.AA_mut_to_DNA_SNP_mut(aa_mut, transcript_id, chrm_path)                          # type: ignore"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv311",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
