{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "f5e2c355-2eda-4702-9886-36badf9a0c49",
   "metadata": {},
   "source": [
    "# Parse Ensembl Gene Annotation Gff3 file and Generate json Files\n",
    "\n",
    "This notebook parses the Ensembl gene annotation GFF3 file, and generates a corresponding json file per gene.\n",
    "\n",
    "The GFF3 file downloaded from [Ensembl gene anotation](https://useast.ensembl.org/Homo_sapiens/Info/Index), by selecting to download GFF3.\n",
    "\n",
    "See [GFF3 format](https://www.ensembl.org/info/website/upload/gff3.html). See also [here](http://ftp.ensembl.org/pub/release-108/gff3/homo_sapiens/README).\n",
    "\n",
    "Note: In general, a protein coding gene may contain non-coding transcripts (RNAs), e.g. lnc_RNA.\n",
    "\n",
    "**To generate the json annotation files:**\n",
    "1. Set the GFF3 file name (`gff3_file` in the `Load the GFF3 File` section below).\n",
    "2. Set the json path (`json_path` in the `Parse GFF3 File and Generate json Files` section below).\n",
    "3. Set `gene_types_to_include` list (in both cells below), which defines the Type of genes to parse. Should be `['gene', 'ncRNA_gene', 'pseudogene']`, or any non-empty subset of it. `'gene'` implies all protein coding genes.\n",
    "4. Run these two sections.\n",
    "\n",
    "**To update to a new Ensembl version follow these steps:**\n",
    "1. Download the GFF3 file (e.g. `Homo_sapiens.GRCh38.109.gff3.gz`, from `https://ftp.ensembl.org/pub/release-109/gff3/homo_sapiens/`) to the path defined by the `gff3_file` variable below (and update `gff3_file` to the new file name).\n",
    "2. Update the `json_path` variable below to the json denstination folder.\n",
    "3. Run this notebook.\n",
    "4. In `Utils/gene_anot_utils.py` update the function `gene2json_file` with the new json denstination folder.\n",
    "\n",
    "**To generate a json file of a gene (or to generate the annotation dictionary of a gene that is written to the json file):**\n",
    "1. Run the `Load the GFF3 File` section below.\n",
    "2. Set the gene name in the `genes` variable in the `Parse GFF3 File and Generate json Files` section below (e.g. `genes = ['EGFR']`).\n",
    "3. Set the other parameters in the `Parse GFF3 File and Generate json Files` section below (e.g., if json file is not needed, set `generate_json_files = False`).\n",
    "4. Run the sections. The variable `gene_anot_s` will then hold the gene annotation dictionary.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "34f76694-d139-4833-a1dc-e93ce269620b",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Global configuration (common to all projects).\n",
    "'''\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os, sys, re\n",
    "import pathlib\n",
    "import pickle\n",
    "import matplotlib.pyplot as plt\n",
    "import glob\n",
    "# import subprocess\n",
    "\n",
    "# make sure PYTHONPATH is set, containing the Utils folder.\n",
    "import mysys_init as si\n",
    "\n",
    "# if sys.platform == 'darwin':\n",
    "#     MACHINE = 'MAC'\n",
    "#     ROOT_path = '/Users/yoramzarai/work/mystuff/Ramot/Projects/Cancer_mut/'\n",
    "#     chromosome_path = pathlib.Path(ROOT_path) / 'Explore_Data' / 'GRCh38_chr'\n",
    "# else:  # Linux TAU machines\n",
    "#     MACHINE = 'POWER'\n",
    "#     ROOT_path = '/tamir2/yoramzar/Projects/Cancer_mut/'\n",
    "#     chromosome_path = pathlib.Path('/tamir2/lab_resources/Genomes/Human/human_hg38/Chromosome')\n",
    "#     #chromosome_path = pathlib.Path(ROOT_path) / 'cancer_proj' / 'human_hg38' / 'Chromosome'\n",
    "    \n",
    "# # adding Utils/ folder to path\n",
    "# utils_path = pathlib.Path(ROOT_path) / 'Utils' \n",
    "# sys.path.append(str(utils_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1b6573d0-a4fb-4d54-9ae7-a4b9f8ffa3cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import ensembl_gene_annotations_utils as egna"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10fe573c-c5f2-4a6f-a3a2-233528dbce83",
   "metadata": {},
   "source": [
    "# Load the GFF3 File\n",
    "\n",
    "The main gene types in the 'Type' column are:\n",
    "1. gene - protein coding genes.\n",
    "2. ncRNA_gene - non-protein coding genes.\n",
    "3. pseudogene\n",
    "\n",
    "See [this](https://useast.ensembl.org/info/genome/genebuild/biotypes.html) for more information.\n",
    "\n",
    "Most of the other values in the 'Type' column are sub-types of these (for example, exon, CDS, lnc_RNA, etc.)\n",
    "\n",
    "For some of the genes, there is no \"Name=\\<name\\>\" string in the Attributes column. This string is available, however, in all protein coding genes. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "64eb56e6-4ac0-473f-b997-d88a2dce9c5a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrm</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>1</td>\n",
       "      <td>248956422</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10469</td>\n",
       "      <td>11240</td>\n",
       "      <td>1.3e+03</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=oe %3D 0.79;logic_name=cpg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10650</td>\n",
       "      <td>10657</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10655</td>\n",
       "      <td>10657</td>\n",
       "      <td>0.999</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>10678</td>\n",
       "      <td>10687</td>\n",
       "      <td>0.999</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>logic_name=eponine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3414006</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>26626966</td>\n",
       "      <td>26627137</td>\n",
       "      <td>0.994</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=rank %3D 1;logic_name=firstef</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3414007</th>\n",
       "      <td>Y</td>\n",
       "      <td>.</td>\n",
       "      <td>biological_region</td>\n",
       "      <td>26627457</td>\n",
       "      <td>26628186</td>\n",
       "      <td>0.997</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>external_name=rank %3D 1;logic_name=firstef</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3414008</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>56855244</td>\n",
       "      <td>56855488</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000235857;Name=CTBP2P1;biotype=p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3414009</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>pseudogenic_transcript</td>\n",
       "      <td>56855244</td>\n",
       "      <td>56855488</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=transcript:ENST00000431853;Parent=gene:ENSG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3414010</th>\n",
       "      <td>Y</td>\n",
       "      <td>havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>56855244</td>\n",
       "      <td>56855488</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=transcript:ENST00000431853;Name=ENSE000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3414011 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chrm  Source                    Type     Start        End    Score  \\\n",
       "0          1  GRCh38              chromosome         1  248956422        .   \n",
       "1          1       .       biological_region     10469      11240  1.3e+03   \n",
       "2          1       .       biological_region     10650      10657    0.999   \n",
       "3          1       .       biological_region     10655      10657    0.999   \n",
       "4          1       .       biological_region     10678      10687    0.999   \n",
       "...      ...     ...                     ...       ...        ...      ...   \n",
       "3414006    Y       .       biological_region  26626966   26627137    0.994   \n",
       "3414007    Y       .       biological_region  26627457   26628186    0.997   \n",
       "3414008    Y  havana              pseudogene  56855244   56855488        .   \n",
       "3414009    Y  havana  pseudogenic_transcript  56855244   56855488        .   \n",
       "3414010    Y  havana                    exon  56855244   56855488        .   \n",
       "\n",
       "        Strand Phase                                         Attributes  \n",
       "0            .     .  ID=chromosome:1;Alias=CM000663.2,chr1,NC_00000...  \n",
       "1            .     .           external_name=oe %3D 0.79;logic_name=cpg  \n",
       "2            +     .                                 logic_name=eponine  \n",
       "3            -     .                                 logic_name=eponine  \n",
       "4            +     .                                 logic_name=eponine  \n",
       "...        ...   ...                                                ...  \n",
       "3414006      -     .        external_name=rank %3D 1;logic_name=firstef  \n",
       "3414007      +     .        external_name=rank %3D 1;logic_name=firstef  \n",
       "3414008      +     .  ID=gene:ENSG00000235857;Name=CTBP2P1;biotype=p...  \n",
       "3414009      +     .  ID=transcript:ENST00000431853;Parent=gene:ENSG...  \n",
       "3414010      +     .  Parent=transcript:ENST00000431853;Name=ENSE000...  \n",
       "\n",
       "[3414011 rows x 9 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Type values:\n",
      "['chromosome', 'biological_region', 'ncRNA_gene', 'lnc_RNA', 'exon', 'pseudogene', 'pseudogenic_transcript', 'miRNA', 'gene', 'mRNA', 'five_prime_UTR', 'CDS', 'three_prime_UTR', 'snRNA', 'transcript', 'ncRNA', 'unconfirmed_transcript', 'snoRNA', 'scRNA', 'rRNA', 'V_gene_segment', 'D_gene_segment', 'J_gene_segment', 'C_gene_segment', 'scaffold', 'tRNA']\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 3414011 entries, 0 to 3414010\n",
      "Data columns (total 9 columns):\n",
      " #   Column      Dtype \n",
      "---  ------      ----- \n",
      " 0   Chrm        object\n",
      " 1   Source      object\n",
      " 2   Type        object\n",
      " 3   Start       int64 \n",
      " 4   End         int64 \n",
      " 5   Score       object\n",
      " 6   Strand      object\n",
      " 7   Phase       object\n",
      " 8   Attributes  object\n",
      "dtypes: int64(2), object(7)\n",
      "memory usage: 234.4+ MB\n"
     ]
    }
   ],
   "source": [
    "gff3_file  = pathlib.Path(si.ROOT_path) / 'Data' / 'Ensembl_gene_annotation' / 'Homo_sapiens.GRCh38.109.gff3'\n",
    "# ============================================================================================================\n",
    "\n",
    "df_anot = pd.read_csv(gff3_file, sep='\\t', comment='#', low_memory=False,\n",
    "                      names=['Chrm', 'Source', 'Type', 'Start', 'End', 'Score', 'Strand', 'Phase', 'Attributes'],\n",
    "                      dtype={\n",
    "                        'Chrm': str,\n",
    "                         'Source': str,\n",
    "                         'Type': str,\n",
    "                         'Start': int,  #  1-based\n",
    "                         'End': int,    #  1-based.\n",
    "                         'Score': str,\n",
    "                         'Strand': str,  # '+' for positive strand, '-' for negative strand.\n",
    "                         'Phase': str,  # some of the rows have non-integer type\n",
    "                         'Attributes': str\n",
    "                     })\n",
    "display(df_anot)\n",
    "\n",
    "# all different types\n",
    "all_types = df_anot['Type'].unique().tolist()\n",
    "print(f\"Type values:\\n{all_types}\")\n",
    "df_anot.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "94958e66-c6e8-4312-a228-d34a43d3eb00",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selecting gene Types...\n",
      "\n",
      "Dataframe containing gene Type:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chrm</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>Source</th>\n",
       "      <th>Type</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Phase</th>\n",
       "      <th>Attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>1</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000186092;Name=OR4F5;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>1</td>\n",
       "      <td>OR4F29</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>450740</td>\n",
       "      <td>451678</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000284733;Name=OR4F29;biotype=pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>502</th>\n",
       "      <td>1</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>685716</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000284662;Name=OR4F16;biotype=pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1097</th>\n",
       "      <td>1</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>923923</td>\n",
       "      <td>944575</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000187634;Name=SAMD11;biotype=pr...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1459</th>\n",
       "      <td>1</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>944203</td>\n",
       "      <td>959309</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000188976;Name=NOC2L;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3412977</th>\n",
       "      <td>Y</td>\n",
       "      <td>BPY2B</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>24607560</td>\n",
       "      <td>24639207</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000183795;Name=BPY2B;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3413043</th>\n",
       "      <td>Y</td>\n",
       "      <td>DAZ3</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>24763069</td>\n",
       "      <td>24813492</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000187191;Name=DAZ3;biotype=prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3413169</th>\n",
       "      <td>Y</td>\n",
       "      <td>DAZ4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>24833843</td>\n",
       "      <td>24907040</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000205916;Name=DAZ4;biotype=prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3413567</th>\n",
       "      <td>Y</td>\n",
       "      <td>BPY2C</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>25030901</td>\n",
       "      <td>25062548</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000185894;Name=BPY2C;biotype=pro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3413759</th>\n",
       "      <td>Y</td>\n",
       "      <td>CDY1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>25622117</td>\n",
       "      <td>25624902</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=gene:ENSG00000172288;Name=CDY1;biotype=prot...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21531 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chrm Gene_name          Source  Type     Start       End Score Strand  \\\n",
       "84         1     OR4F5  ensembl_havana  gene     65419     71585     .      +   \n",
       "370        1    OR4F29  ensembl_havana  gene    450740    451678     .      -   \n",
       "502        1    OR4F16  ensembl_havana  gene    685716    686654     .      -   \n",
       "1097       1    SAMD11  ensembl_havana  gene    923923    944575     .      +   \n",
       "1459       1     NOC2L  ensembl_havana  gene    944203    959309     .      -   \n",
       "...      ...       ...             ...   ...       ...       ...   ...    ...   \n",
       "3412977    Y     BPY2B  ensembl_havana  gene  24607560  24639207     .      +   \n",
       "3413043    Y      DAZ3  ensembl_havana  gene  24763069  24813492     .      -   \n",
       "3413169    Y      DAZ4  ensembl_havana  gene  24833843  24907040     .      +   \n",
       "3413567    Y     BPY2C  ensembl_havana  gene  25030901  25062548     .      -   \n",
       "3413759    Y      CDY1  ensembl_havana  gene  25622117  25624902     .      +   \n",
       "\n",
       "        Phase                                         Attributes  \n",
       "84          .  ID=gene:ENSG00000186092;Name=OR4F5;biotype=pro...  \n",
       "370         .  ID=gene:ENSG00000284733;Name=OR4F29;biotype=pr...  \n",
       "502         .  ID=gene:ENSG00000284662;Name=OR4F16;biotype=pr...  \n",
       "1097        .  ID=gene:ENSG00000187634;Name=SAMD11;biotype=pr...  \n",
       "1459        .  ID=gene:ENSG00000188976;Name=NOC2L;biotype=pro...  \n",
       "...       ...                                                ...  \n",
       "3412977     .  ID=gene:ENSG00000183795;Name=BPY2B;biotype=pro...  \n",
       "3413043     .  ID=gene:ENSG00000187191;Name=DAZ3;biotype=prot...  \n",
       "3413169     .  ID=gene:ENSG00000205916;Name=DAZ4;biotype=prot...  \n",
       "3413567     .  ID=gene:ENSG00000185894;Name=BPY2C;biotype=pro...  \n",
       "3413759     .  ID=gene:ENSG00000172288;Name=CDY1;biotype=prot...  \n",
       "\n",
       "[21531 rows x 10 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 21,519 gene genes.\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Get all required gene names (based on gene_types_to_include variable).\n",
    "\"\"\"\n",
    "# set the gene Type to include. Must be ['gene', 'ncRNA_gene', 'pseudogene'] or any non-empty subset of it\n",
    "gene_types_to_include = ['gene']\n",
    "# =======================================================================================================\n",
    "\n",
    "def get_gene_name(x: pd.Series) -> str:\n",
    "    \"\"\"\n",
    "    If the Attributes has \"Name=<name>\", we return <name>. Otherwise we return <gene_id> \n",
    "    (from the \"gene_id=<gene_id\" string in the Attributes column).\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # using the gene name\n",
    "        return dict([tuple(y.split('=')) for y in x['Attributes'].split(';')])['Name']\n",
    "    except:\n",
    "        # noval transcript (does not have a Name field). Using the gene ID instead. \n",
    "        return dict([tuple(y.split('=')) for y in x['Attributes'].split(';')])['gene_id']\n",
    "# ----------------------------------------------------------------------------------------\n",
    "gene_types = ','.join(gene_types_to_include)\n",
    "print(f\"Selecting {gene_types} Types...\")\n",
    "\n",
    "qry = ''.join([f\"Type == '{gene_types_to_include[0]}'\"] + [f\" or Type == '{g}'\" for g in gene_types_to_include[1:]])\n",
    "#print(f\"\"\"{qry=}\"\"\")\n",
    "df_genes = df_anot.query(qry)\n",
    "    \n",
    "# add gene name column\n",
    "df_genes.insert(1, 'Gene_name', df_genes.apply(get_gene_name, axis=1))\n",
    "print(f\"\\nDataframe containing {gene_types} Type:\")\n",
    "display(df_genes)\n",
    "\n",
    "all_genes = df_genes['Gene_name'].unique().tolist()\n",
    "print(f\"There are {len(all_genes):,} {gene_types} genes.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4337b8c-8363-4e45-9c16-839bd0db9f6e",
   "metadata": {},
   "source": [
    "# Parse GFF3 File and Generate json Files\n",
    "\n",
    "Here we parse the `df_anot` dataframe and generate the json files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7db8aaa8-a9d8-4b68-88a2-bacaf1708d8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsing GFF3 for 21,519 genes of Type ['gene'] (generating json file = True) ...\n",
      "Done 5,000 genes.\n",
      "Done 10,000 genes.\n",
      "Done 15,000 genes.\n",
      "Done 20,000 genes.\n",
      "Done parsing 21,519 genes.\n",
      "CPU times: user 1h 34min 30s, sys: 10min 27s, total: 1h 44min 57s\n",
      "Wall time: 1h 45min\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\"\"\"\n",
    "Parse genes\n",
    "\"\"\"\n",
    "# set a list of gene names to generate json files for. In general, some of the genes do not have \n",
    "# name associated with them, and in this case the gene ID is used as the name. See the cell above.\n",
    "genes = all_genes\n",
    "#genes = ['EGFR', 'IDH1', 'PIK3CA']\n",
    "\n",
    "# this should be identical to the value set above\n",
    "#gene_types_to_include = ['gene']\n",
    "\n",
    "#generate_json_files = False\n",
    "generate_json_files = True\n",
    "\n",
    "# folder to save all json files\n",
    "json_path = pathlib.Path(si.ROOT_path) / 'Ensembl_gene_anot' / 'Ensembl_109_gene_annotations' if si.MACHINE=='MAC' else \\\n",
    "pathlib.Path('/tamir2/lab_resources/Genomes/Human/human_hg38/Ensembl_109_gene_annotations')\n",
    "# =============================================================================================================================================================================================\n",
    "\n",
    "# All Type values for a gene. This is for verification only.\n",
    "gene_type_values = ['gene', 'ncRNA_gene', 'pseudogene']\n",
    "\n",
    "# gene's transcript types to process\n",
    "types_transcript_processed = ['mRNA', 'lnc_RNA', 'pseudogenic_transcript', \n",
    "                              'ncRNA', 'miRNA', 'snoRNA',\n",
    "                              'snRNA', 'scRNA', 'rRNA', 'tRNA']\n",
    "\n",
    "# sub-transcript types to process\n",
    "types_in_transcript = ['CDS', 'exon', 'five_prime_UTR', 'three_prime_UTR']\n",
    "# --------------------------------------------------------------------------\n",
    "\n",
    "print(f\"Parsing GFF3 for {len(genes):,} genes of Type {gene_types_to_include} (generating json file = {generate_json_files}) ...\")\n",
    "\n",
    "if generate_json_files:\n",
    "    json_path.mkdir(exist_ok=True)\n",
    "\n",
    "for i, gene in enumerate(genes, start=1):\n",
    "    if i % 5000 == 0:\n",
    "        print(f\"Done {i:,} genes.\")\n",
    "\n",
    "    # generate a dataframe starting from the gene of interest (the dataframe may contain other genes as well, which will be ignored)\n",
    "    if (df_gene_all := egna.get_df_start_with_gene(df_anot, gene, gene_types_to_include)) is None:\n",
    "        continue\n",
    "        \n",
    "    # verify\n",
    "    first_row = df_gene_all.iloc[0]\n",
    "    if first_row['Type'] not in gene_type_values:\n",
    "        print(f\"df_gene_all for {gene=} does not start with a row with a Type in {gene_type_values} !!\")\n",
    "        continue\n",
    "        \n",
    "    d = dict([tuple(x.split('=')) for x in first_row['Attributes'].split(';')])\n",
    "    try:\n",
    "        first_row_gene = d['Name']\n",
    "    except KeyError:\n",
    "        # non-coding genes have no \"Name\" field. In these cases we use the gene ID as the name.\n",
    "        first_row_gene = d['gene_id']\n",
    "        \n",
    "    if gene != first_row_gene:\n",
    "        print(f\"{first_row_gene=} differ from {gene=} !!\")\n",
    "        continue\n",
    "    \n",
    "    # generate the annotation dictionary\n",
    "    if (gene_anot_s := egna.parse_ensembl_gene_anot_gene(df_gene_all, types_transcript_processed, types_in_transcript)) == {}:\n",
    "        print(f\"Error in annotating {gene=} !!\")\n",
    "    else:\n",
    "        if generate_json_files:\n",
    "            with open(json_path / f\"{gene}_annotation.json\", 'w') as fp:\n",
    "                json.dump(gene_anot_s, fp)\n",
    "    \n",
    "print(f\"Done parsing {len(genes):,} genes.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b24a01b9-4efb-4a36-9603-12a132232943",
   "metadata": {},
   "outputs": [],
   "source": [
    "# display annotation for the last gene\n",
    "display(df_gene_all)\n",
    "egna.display_gene_annotation_structure(gene_anot_s)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17afbdd8-17c3-4378-8c0e-d7effaf4c031",
   "metadata": {},
   "source": [
    "# ScratchPad\n",
    "Ignore this section."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b8d6d1f-f567-4360-abc9-00b25043781a",
   "metadata": {},
   "source": [
    "## Testing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4869b42-d3b7-4fd9-ba64-db7b248174a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import mRNA_utils as mrnau\n",
    "import ensembl_rest_utils as erst\n",
    "import translation as tran\n",
    "\n",
    "get_chrm_file_name = mrnau.get_chrm_file_name_mac if MACHINE=='MAC' else mrnau.get_chrm_file_name_power\n",
    "mrna_data_path = pathlib.Path(ROOT_path) / 'Explore_Splice' / 'mrna_data' if MACHINE=='MAC' else pathlib.Path('/tamir2/lab_resources/Genomes/Human/human_hg38/mRNAsAll_June2021/')\n",
    "\n",
    "report_path = pathlib.Path(ROOT_path) / 'Ensembl_gene_anot' / 'Reports'\n",
    "\n",
    "report_path.mkdir(exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "937403f8-6787-473c-813b-f9eb53594f8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "compare_with_REST = True\n",
    "indx = 1 #1 # pick transcript index to test\n",
    "# ========================================\n",
    "\n",
    "def get_sequence_from_start_end_segments(start: list, end: list, rev: bool, chrm_path: str, offset: int=0) -> str:\n",
    "    \"\"\"\n",
    "    Get the chromosome sequence, at the correponding strand (based on rev input) from \n",
    "    the concatenation of segments defined by start and end list.\n",
    "    \n",
    "    rev == True implies start[i]>end[i], start[j]<start[i], end[j]<end[i], for j>i, otherwise \n",
    "                        start[i]<end[i], start[j]>start[i], end[j]>end[i], for j>1. \n",
    "    \"\"\"\n",
    "    # convert to increasing lists\n",
    "    a_start, a_end = (end[::-1], start[::-1]) if rev else (start, end)\n",
    "    seq = ''.join([mrnau.pull_fasta_seq(chrm_path, a_s, 0, a_e - a_s, False)\n",
    "                   for a_s, a_e in zip(a_start, a_end)])\n",
    "    return tran.reverse_complement(seq)[offset:] if rev else seq[offset:]\n",
    "# ========================================================================================\n",
    "\n",
    "\n",
    "test_transcript = gene_anot_s['transcripts'][indx]\n",
    "print(f\"Transcript=\\n{test_transcript}\")\n",
    "\n",
    "if test_transcript['transcript_biotype'] != 'protein_coding':\n",
    "    raise Exception(f\"Exception: transcript{indx} ({test_transcript['transcript_name']}) corresponds to a non protein coding transcript !!\")\n",
    "\n",
    "\n",
    "chrm = f\"chr{gene_anot_s['chrm']}\"\n",
    "rev = gene_anot_s['rev']\n",
    "transcript_id = test_transcript['transcript_id']\n",
    "transcript_name = test_transcript['transcript_name']\n",
    "p_id = test_transcript['protein_id']\n",
    "\n",
    "cds_offset = test_transcript['CDS_phase'][0]\n",
    "\n",
    "print(f\"\\n{gene=}, {transcript_id=}, {p_id=}, {transcript_name=}, {chrm=}, {rev=}\")\n",
    "\n",
    "if compare_with_REST:\n",
    "    ensb_trans = erst.lookup_id(transcript_id)\n",
    "    ensb_pid = ensb_trans['Translation']['id']\n",
    "    print(f\"\\tREST: {transcript_id=} ==> {ensb_pid=}\")\n",
    "    assert ensb_pid == p_id, f\"{p_id=} and {ensb_pid=} mismatch !!\"\n",
    "    ensb_aa = erst.sequence_endpoint_base(ensb_pid)\n",
    "    \n",
    "    # need to check this more carefully\n",
    "    if cds_offset != 0:\n",
    "        ensb_aa = ensb_aa[1:]\n",
    "\n",
    "\n",
    "# comparing with Transcripts in mRNA_utils\n",
    "trans_obj = mrnau.Transcripts(gene)\n",
    "chrm_path = chromosome_path / get_chrm_file_name(trans_obj.chrm)\n",
    "aa = trans_obj.AA(transcript_id, chrm_path)\n",
    "if aa is not None:\n",
    "    aa_n = aa[:-1] if aa[-1] == '?' else aa\n",
    "\n",
    "\n",
    "    # compare with REST\n",
    "    if compare_with_REST:\n",
    "        if ensb_aa == aa_n:\n",
    "            print(f\"\\tAA matches REST AA.\")\n",
    "        else:\n",
    "            print(f\"AA and REST do not agree !!\")\n",
    "    else:\n",
    "        print(f\"Comparing with REST is set to OFF !\")\n",
    "    \n",
    "\"\"\"\n",
    "Get AA from gene_anot_s\n",
    "\"\"\"\n",
    "ORF = get_sequence_from_start_end_segments(test_transcript['CDS_start'], test_transcript['CDS_end'], rev, chrm_path, offset=cds_offset)\n",
    "# this includes the stop codon\n",
    "aa_s = tran.convert_aa_32IUPAC(tran.translate(ORF))\n",
    "\n",
    "aa_s_n = aa_s[:-1] if aa_s[-1] == '?' else aa_s\n",
    "\n",
    "if aa is not None:\n",
    "    if aa_s_n == aa_n:\n",
    "        print(f\"\\nAA matches with gene_anot_s\")\n",
    "    else:\n",
    "        print(f\"\\nMissmatch between gene_anot_s and mrnau !!!!!\")\n",
    "    \n",
    "if compare_with_REST:\n",
    "    if aa_s_n == ensb_aa:\n",
    "        print(f\"\\nREST matches with gene_anot_s\")\n",
    "    else:\n",
    "        print(f\"\\nMissmatch between gene_anot_s and REST !!!!!\")\n",
    "    \n",
    "\"\"\"\n",
    "Get mRNA (transcript) sequence\n",
    "\"\"\"\n",
    "mRNA_seq = get_sequence_from_start_end_segments(test_transcript['exon_start'], test_transcript['exon_end'], rev, chrm_path)\n",
    "\n",
    "\"\"\"\n",
    "Get 5'UTR\n",
    "\"\"\"\n",
    "if test_transcript['5UTR_start'] == []:\n",
    "    print(\"No 5UTR\")\n",
    "else:\n",
    "    UTR5_seq = get_sequence_from_start_end_segments(test_transcript['5UTR_start'], test_transcript['5UTR_end'], rev, chrm_path)\n",
    "    print(f\"\\n5UTR=\\n{UTR5_seq}\")\n",
    "\n",
    "\"\"\"\n",
    "Get 3UTR\n",
    "\"\"\"\n",
    "if test_transcript['3UTR_start'] == []:\n",
    "    print(\"No 3UTR\")\n",
    "else:\n",
    "    UTR3_seq = get_sequence_from_start_end_segments(test_transcript['3UTR_start'], test_transcript['3UTR_end'], rev, chrm_path)\n",
    "    print(f\"\\n3UTR=\\n{UTR3_seq}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9fee752-a1a6-4242-9dce-e94ebbb52a50",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(aa_s_n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cb46ca8-0f88-4281-bee2-f643429b10ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(ensb_aa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5e20bd0-93f0-4dd8-badc-4868b6caa7a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(ORF[:10])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv311",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  },
  "vscode": {
   "interpreter": {
    "hash": "aab3f81e15692bda916e0752fee5273bf94e969e803fd9abfc8404476062fb6b"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
